<html><head><meta http-equiv="Content-Type" content="text/html; charset=ISO-8859-1"><title>BAli-Phy User's Guide v2.2.2</title><link rel="stylesheet" type="text/css" href="docbook.css"><meta name="generator" content="DocBook XSL Stylesheets V1.78.1"></head><body bgcolor="white" text="black" link="#0000FF" vlink="#840084" alink="#0000FF"><div lang="en" class="article"><div class="titlepage"><div><div><h2 class="title"><a name="idp57580048"></a><span class="application">BAli-Phy</span> User's Guide v2.2.2</h2></div><div><div class="author"><h3 class="author"><span class="firstname">Benjamin</span> <span class="surname">Redelings</span></h3></div></div></div><hr></div><div class="toc"><p><b>Table of Contents</b></p><dl class="toc"><dt><span class="section"><a href="#intro">1. Introduction</a></span></dt><dt><span class="section"><a href="#installation">2. Installation</a></span></dt><dd><dl><dt><span class="section"><a href="#pre-requisites">2.1. Hardware requirements</a></span></dt><dt><span class="section"><a href="#idp57598896">2.2. MS Windows requirements</a></span></dt><dt><span class="section"><a href="#idp57611104">2.3. Mac OS X requirements</a></span></dt><dt><span class="section"><a href="#idp57614672">2.4. Linux requirements</a></span></dt><dt><span class="section"><a href="#idp57616368">2.5. Installing BAli-Phy executables</a></span></dt><dt><span class="section"><a href="#idp57639520">2.6. Adding BAli-Phy to your <code class="envar">PATH</code></a></span></dt><dt><span class="section"><a href="#software_req">2.7. Additional required software</a></span></dt><dt><span class="section"><a href="#tests">2.8. Testing the installed software</a></span></dt></dl></dd><dt><span class="section"><a href="#compilation">3. Compiling <span class="application">BAli-Phy</span></a></span></dt><dd><dl><dt><span class="section"><a href="#idp57679056">3.1. Software requirements</a></span></dt><dt><span class="section"><a href="#idp57695152">3.2. Quick Start</a></span></dt><dt><span class="section"><a href="#idp57703968">3.3. Options to <span class="command"><strong>configure</strong></span></a></span></dt><dt><span class="section"><a href="#gitsource">3.4. Generating the <span class="command"><strong>configure</strong></span> script
	and Makefiles (<span class="application">git</span> only)</a></span></dt><dt><span class="section"><a href="#idp57735632">3.5. Installing when compiling from source</a></span></dt></dl></dd><dt><span class="section"><a href="#running">4. Running the program</a></span></dt><dd><dl><dt><span class="section"><a href="#idp57742800">4.1. Quick Start</a></span></dt><dt><span class="section"><a href="#idp57751696">4.2. Command line options</a></span></dt><dt><span class="section"><a href="#idp57759248">4.3. Multiple genes or data partitions</a></span></dt><dt><span class="section"><a href="#idp57778160">4.4. Option files (Scripts)</a></span></dt><dt><span class="section"><a href="#examples">4.5. Examples</a></span></dt><dt><span class="section"><a href="#command_line_options">4.6. Command-line options: An overview</a></span></dt></dl></dd><dt><span class="section"><a href="#input">5. Input</a></span></dt><dd><dl><dt><span class="section"><a href="#idp57854288">5.1. Sequence formats</a></span></dt><dt><span class="section"><a href="#idp57860064">5.2. Is my data set too large?</a></span></dt></dl></dd><dt><span class="section"><a href="#output">6. Output</a></span></dt><dd><dl><dt><span class="section"><a href="#idp57876000">6.1. Output directory</a></span></dt><dt><span class="section"><a href="#idp57881168">6.2. Output options</a></span></dt><dt><span class="section"><a href="#idp57883616">6.3. Output files</a></span></dt><dt><span class="section"><a href="#idp57943552">6.4. Summarizing the output</a></span></dt><dt><span class="section"><a href="#analysis">6.5. Summarizing the output - scripted</a></span></dt></dl></dd><dt><span class="section"><a href="#subst_models">7. Substitution models</a></span></dt><dd><dl><dt><span class="section"><a href="#idp58044000">7.1. Default substitution models</a></span></dt><dt><span class="section"><a href="#subst_basic">7.2. Basic CTMC models</a></span></dt><dt><span class="section"><a href="#subst_freq">7.3. Substitution Frequency models</a></span></dt><dt><span class="section"><a href="#idp58274288">7.4. Substitution Mixture Models</a></span></dt><dt><span class="section"><a href="#idp58362320">7.5. The branch-site substitution model</a></span></dt><dt><span class="section"><a href="#idp58372192">7.6. Substitution model examples</a></span></dt><dt><span class="section"><a href="#genetic-codes">7.7. Genetic Codes</a></span></dt></dl></dd><dt><span class="section"><a href="#indel_models">8. Insertion/deletion models</a></span></dt><dt><span class="section"><a href="#alignment_constraints">9. Alignment constraints</a></span></dt><dd><dl><dt><span class="section"><a href="#idp58424528">9.1. Syntax</a></span></dt><dt><span class="section"><a href="#idp58427312">9.2. Examples</a></span></dt><dt><span class="section"><a href="#idp58431584">9.3. Computing the constraints</a></span></dt></dl></dd><dt><span class="section"><a href="#mixing_and_convergence">10. Convergence and Mixing: Is it done yet?</a></span></dt><dd><dl><dt><span class="section"><a href="#idp58439968">10.1. Definition of Convergence</a></span></dt><dt><span class="section"><a href="#idp58442256">10.2. Definition of Mixing</a></span></dt><dt><span class="section"><a href="#idp58444272">10.3. Diagnostics and multiple independent chains</a></span></dt><dt><span class="section"><a href="#idp58449872">10.4. Diagnostics: Variation in split frequencies across runs (ASDSF/MSDSF)</a></span></dt><dt><span class="section"><a href="#idp58459648">10.5. Diagnostics: Potential Scale Reduction Factors (PSRF)</a></span></dt><dt><span class="section"><a href="#idp58465168">10.6. Diagnostics: Effective sample sizes (ESS)</a></span></dt><dt><span class="section"><a href="#idp58474160">10.7. Diagnostics: Stabilization</a></span></dt></dl></dd><dt><span class="section"><a href="#auxiliary_tools">11. Auxiliary tools</a></span></dt><dd><dl><dt><span class="section"><a href="#idp58497504">11.1. alignment-find</a></span></dt><dt><span class="section"><a href="#idp58499664">11.2. alignment-draw</a></span></dt><dt><span class="section"><a href="#idp58502736">11.3. alignment-thin</a></span></dt><dt><span class="section"><a href="#idp58506000">11.4. alignment-chop-internal</a></span></dt><dt><span class="section"><a href="#idp58508464">11.5. alignment-info</a></span></dt><dt><span class="section"><a href="#idp58512048">11.6. alignment-indices</a></span></dt><dt><span class="section"><a href="#idp58515312">11.7. alignment-cat</a></span></dt><dt><span class="section"><a href="#idp58518352">11.8. trees-consensus</a></span></dt><dt><span class="section"><a href="#idp58521728">11.9. trees-bootstrap</a></span></dt><dt><span class="section"><a href="#idp58527088">11.10. trees-to-SRQ</a></span></dt></dl></dd><dt><span class="section"><a href="#FAQ">12. Frequently Asked Questions (FAQ)</a></span></dt><dd><dl><dt><span class="section"><a href="#idp58534000">12.1. Input files</a></span></dt><dt><span class="section"><a href="#idp58545728">12.2. Running <span class="command"><strong>bali-phy</strong></span>.</a></span></dt><dt><span class="section"><a href="#idp58562960">12.3. Run-time error messages</a></span></dt><dt><span class="section"><a href="#idp58568992">12.4. Stopping <span class="command"><strong>bali-phy</strong></span>.</a></span></dt><dt><span class="section"><a href="#idp58593248">12.5. Interpreting the results.</a></span></dt><dt><span class="section"><a href="#idp58600656">12.6. How do I...</a></span></dt></dl></dd></dl></div>
  

  <div class="section"><div class="titlepage"><div><div><h2 class="title" style="clear: both"><a name="intro"></a>1. Introduction</h2></div></div></div>
    <p><span class="application">BAli-Phy</span> is a Unix command
      line program that is developed primarily on
      Linux.  <span class="application">BAli-Phy</span> also runs on
      Windows and Mac OS X, but it is not a GUI program and so you
      must run it in a terminal.  Therefore, you might want to keep a
      <a class="ulink" href="http://www.ee.surrey.ac.uk/Teaching/Unix" target="_top">Unix
      tutorial</a> or <a class="ulink" href="http://www.rain.org/~mkummel/unix.html" target="_top">Unix
      cheat sheet</a> handy while you work.
    </p>
  </div>

  <div class="section"><div class="titlepage"><div><div><h2 class="title" style="clear: both"><a name="installation"></a>2. Installation</h2></div></div></div>

  <p>You can install BAli-Phy by downloading compiled executables from the <a class="ulink" href="http://www.bali-phy.org/download.php" target="_top">website</a>.</p>
  <p>On OSX 10.8 or higher, you could also install with <a class="ulink" href="http://brew.sh/" target="_top">homebrew</a> if you have the XCode 5 compiler.  The recipe is in <a class="ulink" href="https://github.com/Homebrew/homebrew-science/blob/master/README.md" target="_top">homebrew/science</a>.</p>
  <div class="section"><div class="titlepage"><div><div><h3 class="title"><a name="pre-requisites"></a>2.1. Hardware requirements</h3></div></div></div>
    
    <p>We recommend running BAli-Phy on a computing cluster for long runs.
    A computing cluster can speed up the analysis by allowing
    you to run several identical MCMC chains simultaneously and then
    pool the resulting samples.  You also don't need to worry that
    logging out or turning off the computer will terminate the run
    early.  Result files can be copied back to a laptop or desktop for
    viewing.</p> 

    <p>
      We typically run <span class="application">BAli-Phy</span> on Core2
      processors with 8Gb of RAM.  You may need a 64-bit executable and a 64-bit
      version of your operating system to be able to analyze large data sets that
      consume more that 2Gb of RAM.
    </p>

    <p>
    </p>
  </div>

    <div class="section"><div class="titlepage"><div><div><h3 class="title"><a name="idp57598896"></a>2.2. MS Windows requirements</h3></div></div></div>
      <p>Before you can use <span class="application">BAli-Phy</span> on
      MS Windows, you must first install <a class="ulink" href="http://www.cygwin.com/install.html" target="_top">Cygwin</a>.
      Cygwin is a Unix/Linux command-line environment for Windows.
While running the Cygwin installer
	<code class="filename">setup.exe</code>, you will be given an
	opportunity to select additional packages.
	</p><div class="itemizedlist"><ul class="itemizedlist" style="list-style-type: disc; "><li class="listitem"><p>From <span class="guilabel">Interpreters</span>, select <span class="guilabel">perl</span>.</p></li><li class="listitem"><p>From <span class="guilabel">Web</span>, select <span class="guilabel">wget</span>.</p></li><li class="listitem"><p>From <span class="guilabel">Editors</span>, select <span class="guilabel">nano</span>.</p></li></ul></div><p>
	
      </p>

      <p>You may then access the Unix command line environment by running the
      <span class="application">Cygwin Terminal</span> application (not the
      normal windows command line).</p> 

      <p>You might wish to save the installer on your
      desktop in case you want to run it again, since you can use it
      to install additional packages later.</p>

      <p>In order to access your Windows files from inside cygwin, note
      that <code class="filename">C:/</code> is accessed as
      <code class="filename">/cygdrive/c/</code>.  Thus, for example, you can
      access downloaded files at <code class="filename">/cygdrive/c/Documents\ and\
      Settings/<em class="replaceable"><code>username</code></em>/Downloads/</code>.
      </p>
    </div>

    <div class="section"><div class="titlepage"><div><div><h3 class="title"><a name="idp57611104"></a>2.3. Mac OS X requirements</h3></div></div></div>
    <p>We recommend Mac OS X version 10.6 (or higher).</p>
    <p>You can now install with <a class="ulink" href="http://brew.sh/" target="_top">homebrew</a> if you have the XCode 5 compiler.  The recipe is in <a class="ulink" href="https://github.com/Homebrew/homebrew-science/blob/master/README.md" target="_top">homebrew/science</a>.</p>
    </div>

    <div class="section"><div class="titlepage"><div><div><h3 class="title"><a name="idp57614672"></a>2.4. Linux requirements</h3></div></div></div>
      <p>No extra requirements.</p>
    </div>

    <div class="section"><div class="titlepage"><div><div><h3 class="title"><a name="idp57616368"></a>2.5. Installing BAli-Phy executables</h3></div></div></div>

    <div class="orderedlist"><ol class="orderedlist" type="1"><li class="listitem"><p>Start by opening a Unix terminal window.  (On Apple, this is the <span class="application">Terminal</span>
    application; on Windows it will be the <span class="application">Cygwin Terminal</span>, not the
    Windows command prompt.)</p></li><li class="listitem"><p>Make a directory called <code class="filename">Applications</code> in your
    home directory:
</p><pre class="screen"><code class="prompt">%</code> <strong class="userinput"><code>mkdir ~/Applications</code></strong></pre><p>
    </p><div class="note" style="margin-left: 0.5in; margin-right: 0.5in;"><h3 class="title">Note</h3>
    Note that <strong class="userinput"><code>~</code></strong> is a synonym for
    <strong class="userinput"><code>$HOME</code></strong>, your home directory.  You can
    find out what your home directory is by typing
<pre class="screen"><code class="prompt">%</code> <strong class="userinput"><code>echo $HOME</code></strong></pre>
or
<pre class="screen"><code class="prompt">%</code> <strong class="userinput"><code>echo ~</code></strong></pre></div><p>
</p></li><li class="listitem"><p><a class="ulink" href="http://www.bali-phy.org/download.php" target="_top">Download</a> 
BAli-Phy executables for your operating system from the web site using your browser.  Choose the 64-bit executables unless you're operating system can only run 32-bit executables.  Save them to the <code class="filename">~/Applications</code> directory that you just created on the command line. Then check to see that the file is there:
</p><pre class="screen"><code class="prompt">%</code> <strong class="userinput"><code>ls ~/Applications</code></strong>
bali-phy-2.2.2.tgz
</pre><p>
</p>
<p>Alternatively, you can download the file directly from the command line using <strong class="userinput"><code>wget</code></strong> if you know the URL:
</p><pre class="screen"><code class="prompt">%</code> <strong class="userinput"><code>cd ~/Applications</code></strong>
<code class="prompt">%</code> <strong class="userinput"><code>wget http://www.bali-phy.org/bali-phy-2.2.2.tar.gz</code></strong>
<code class="prompt">%</code> <strong class="userinput"><code>ls</code></strong>
bali-phy-2.2.2.tgz
</pre><p>
</p>
</li><li class="listitem"><p>Extract the compressed archive on the Unix (or
    Cygwin) command line using the <span class="command"><strong>tar</strong></span> command:
</p><pre class="screen"><code class="prompt">%</code> <strong class="userinput"><code>cd ~/Applications</code></strong>
<code class="prompt">%</code> <strong class="userinput"><code>tar -zxf bali-phy-2.2.2.tgz</code></strong>
<code class="prompt">%</code> <strong class="userinput"><code>ls</code></strong>
bali-phy-2.2.2.tgz  bali-phy-2.2.2/
</pre><p>
    </p></li><li class="listitem"><p>Finally, test that the program can be run.
</p><pre class="screen"><code class="prompt">%</code> <strong class="userinput"><code>~/Applications/bali-phy-2.2.2/bin/bali-phy -v</code></strong>
VERSION: 2.2.0  [2.2 commit 814e5134+]  (Dec 20 2013 20:24:02)
BUILD: Dec 20 2013 20:28:12
ARCH: x86_64-unknown-linux-gnu
COMPILER: GCC 4.8.2
FLAGS: -isystem $(top_srcdir)/boost/include -ffast-math -DNDEBUG -DNDEBUG_DP -funroll-loops -std=c++11 -finline-limit=1000   -pipe -O3
</pre><p>
</p></li></ol></div>
    </div>      

    <div class="section"><div class="titlepage"><div><div><h3 class="title"><a name="idp57639520"></a>2.6. Adding BAli-Phy to your <code class="envar">PATH</code></h3></div></div></div>

      <div class="section"><div class="titlepage"><div><div><h4 class="title"><a name="idp57639776"></a>2.6.1. I have a path?</h4></div></div></div>
      <p>
	If you installed <span class="application">BAli-Phy</span> to the directory
	<code class="filename">~/Applications</code>, then you can run
	bali-phy by typing <span class="command"><strong>~/Applications/bali-phy-2.2.2/bin/bali-phy</strong></span>.
	However, it would be much nicer to simply type
	<span class="command"><strong>bali-phy</strong></span> and let the computer find the
	executable for you.  This can be achieved by putting the directory
	that contains the <span class="application">BAli-Phy</span> executables into
	your "path".  	The "path" is a colon-separated list of directories that is
	searched to find program names that you type.  It is stored in an
	environment variable called <code class="envar">PATH</code>.
	</p>
      <p>
	Setting your <code class="envar">PATH</code> is also a pre-requisite for running
	the <span class="command"><strong>bp-analyze.pl</strong></span> script to summarize your
	MCMC runs.
      </p>
      </div>

      <div class="section"><div class="titlepage"><div><div><h4 class="title"><a name="idp57646336"></a>2.6.2. Examining your <code class="envar">PATH</code></h4></div></div></div>
      <p>
	You can examine the current value of
	this environment variable by typing:
	</p><pre class="screen"><code class="prompt">%</code> <strong class="userinput"><code>echo $PATH</code></strong></pre><p>
	We will assume that you extracted the bali-phy archive in
	<code class="filename">~/Applications</code> and so you want to add
	<code class="filename">$HOME/Applications/bali-phy-2.2.2/bin</code>
	to your <code class="envar">PATH</code>.  (If you installed to another directory,
	replace <code class="filename">$HOME/Applications/bali-phy-2.2.2/</code> with that directory.)
      </p>
      </div>

      <div class="section"><div class="titlepage"><div><div><h4 class="title"><a name="idp57650096"></a>2.6.3. Adding BAli-Phy to your <code class="envar">PATH</code></h4></div></div></div>
      <p>The commands
	for doing this depend on what "shell" you are using.  Type
	<span class="command"><strong>echo $SHELL</strong></span> to find out. If your
	shell is <span class="command"><strong>sh</strong></span> or 
	<span class="command"><strong>bash</strong></span> then the command looks like this: 
	</p><pre class="screen"><code class="prompt">%</code> <strong class="userinput"><code>PATH=$HOME/Applications/bali-phy-2.2.2/bin:$PATH</code></strong></pre><p>
	If your shell is <span class="command"><strong>csh</strong></span> or
	<span class="command"><strong>tcsh</strong></span>, then the command looks like this:
	</p><pre class="screen"><code class="prompt">%</code> <strong class="userinput"><code>setenv PATH $HOME/Applications/bali-phy-2.2.2/bin:$PATH</code></strong></pre><p>
	Note that these commands will only affect the window you are typing
	in, and will vanish when you reboot.   
      </p>
      </div>

      <div class="section"><div class="titlepage"><div><div><h4 class="title"><a name="idp57655424"></a>2.6.4. Making the change stick</h4></div></div></div>
	<p>
	  To make this change survives when you logout or reboot, open
	  your shell configuration file in a text editor, and add the
	  command on a line by itself.  This will ensure that it is
	  run every time you log in.
	</p>

	<p>To find the right configuration file, look in your $HOME directory
	  for <code class="filename">.profile</code> (for the Bourne shell <span class="command"><strong>sh</strong></span>), 
	  <code class="filename">.bash_profile</code> (for BASH), or
	  <code class="filename">.login</code> (for tcsh).  You may have to
	  create the file if it is not present.  On Cygwin, you should
	  put the change in the file <code class="filename">.bashrc</code>.
	</p>

	<p>If you do not know which directory is your home
	directory, you can find its full name by typing:
	</p><pre class="screen"><code class="prompt">%</code> <strong class="userinput"><code>echo $HOME</code></strong></pre><p>
	</p>
      </div>
    </div>

    <div class="section"><div class="titlepage"><div><div><h3 class="title"><a name="software_req"></a>2.7. Additional required software</h3></div></div></div>

    <p>The following software is important to install:
    </p><div class="itemizedlist"><ul class="itemizedlist" style="list-style-type: disc; "><li class="listitem"><p>The graphical MCMC diagnostic program
      <a class="ulink" href="http://tree.bio.ed.ac.uk/software/tracer/" target="_top">Tracer</a>.
      </p></li></ul></div><p>
    </p>

    <p>The following software is recommended to install:

    </p><div class="itemizedlist"><ul class="itemizedlist" style="list-style-type: disc; "><li class="listitem"><p>The plotting program gnuplot.</p></li><li class="listitem"><p>The phylogeny-viewer <a class="ulink" href="http://tree.bio.ed.ac.uk/software/figtree/" target="_top">FigTree</a>.</p></li><li class="listitem"><p>The alignment-viewer <a class="ulink" href="http://pbil.univ-lyon1.fr/software/seaview.html" target="_top">Seaview</a>.</p></li></ul></div><p>
    GNUplot can be installed using the Cygwin installer on Windows
    systems.  It can also be installed using <a class="ulink" href="http://www.macports.org" target="_top">macports</a> or
    <a class="ulink" href="http://http://mxcl.github.com/homebrew/" target="_top">homebrew</a>
    on Macintosh systems, but installing these package managers
    requires first installing Xcode Developer Tools, which requires an Apple Developer
    ID, and is not trivial.
    </p>

    <p>
      For those who wish to install macports, a <span class="application">bali-phy</span> package is available.
    </p>
      
    </div>

    <div class="section"><div class="titlepage"><div><div><h3 class="title"><a name="tests"></a>2.8. Testing the installed software</h3></div></div></div>
    <p>In order to determine that the software has been correctly installed, and the <code class="envar">PATH</code> has been correctly set, run the following commands:
</p><pre class="screen"><code class="prompt">%</code> <strong class="userinput"><code>bali-phy ~/Applications/bali-phy-2.2.2/share/bali-phy/examples/5S-rRNA/5d.fasta --iter=50</code></strong>
<code class="prompt">%</code> <strong class="userinput"><code>bp-analyze.pl 5d-1/</code></strong></pre><p>
    </p>
    <p>Furthermore, the directory <code class="filename">5d-1</code> should contain a file called <code class="filename">C1.p</code>.  You should be able to load this file in Tracer, although the chain will not really have converged yet.</p>
    </div>

  </div>

  <div class="section"><div class="titlepage"><div><div><h2 class="title" style="clear: both"><a name="compilation"></a>3. Compiling <span class="application">BAli-Phy</span></h2></div></div></div>
    

    <p>Most users will not need to compile <span class="application">BAli-Phy</span>
      and can skip this section, because they can use the precompiled 
      executables from the official website for Linux, Mac, and
      Windows.  However, compiling <span class="application">BAli-Phy</span>
      is intended to be a relatively painless process.
    </p>

    <p>If you are compiling "live" source
      code that you checked out using GIT (and you probably
      aren't) then you need to follow the directions in <a class="xref" href="#gitsource" title="3.4. Generating the configure script and Makefiles (git only)">Section 3.4, &#8220;Generating the <span class="command"><strong>configure</strong></span> script
	and Makefiles (<span class="application">git</span> only)&#8221;</a> before you start compiling.
    </p>

    <div class="section"><div class="titlepage"><div><div><h3 class="title"><a name="idp57679056"></a>3.1. Software requirements</h3></div></div></div>
      

    <p>In order to compile BAli-Phy, you need a C++ compiler than can understands the C++11 standard.  </p>
    <div class="section"><div class="titlepage"><div><div><h4 class="title"><a name="idp57680624"></a>3.1.1. Compilers: Linux</h4></div></div></div>
    <p>On Linux we recommend the GNU C++ Compiler (<a class="ulink" href="http://gcc.gnu.org" target="_top">GCC</a>) version 4.8 (or higher).
    </p>
    <p>You may also use the <a class="ulink" href="http://clang.llvm.org" target="_top">Clang</a> compiler version 3.4.0 or higher.
    </p>
    </div>

    <div class="section"><div class="titlepage"><div><div><h4 class="title"><a name="idp57683664"></a>3.1.2. Compilers: Mac</h4></div></div></div>
    <p>On Mac 10.8 and above, the simplest method is to install <a class="ulink" href="https://developer.apple.com/xcode/" target="_top">XCode</a> version 5.0 (or newer).  Then you can use the default compiler (<a class="ulink" href="http://clang.llvm.org" target="_top">Clang</a>).
</p><div class="note" style="margin-left: 0.5in; margin-right: 0.5in;"><h3 class="title">Note</h3>Note that you must add the option <strong class="userinput"><code>CXXFLAGS=-ftemplate-depth=256</code></strong> to the configure script when using clang.  The default template depth of 128 is too low.  See below on how to add this option.
    </div><p>
    </p>
    
    <p>You can now install with <a class="ulink" href="http://brew.sh/" target="_top">homebrew</a> if you have XCode 5.  The recipe is in <a class="ulink" href="https://github.com/Homebrew/homebrew-science/blob/master/README.md" target="_top">homebrew/science</a>.</p>

    </div>
    
    <div class="section"><div class="titlepage"><div><div><h4 class="title"><a name="idp57688864"></a>3.1.3. Compilers: Windows</h4></div></div></div>
    <p>On Windows we recommended the mingw-64 GCC 4.8 compiler.  This compiler can build both 32-bit and 64-bit executables.  You can install it through the compiler through <a class="ulink" href="http://www.cygwin.com/" target="_top">Cygwin</a>.  The compiled executables can be run anywhere, and do not need Cygwin to run.
    </p>
    </div>

    <div class="section"><div class="titlepage"><div><div><h4 class="title"><a name="idp57691072"></a>3.1.4. Optional software libraries</h4></div></div></div>
    <p>This is not strictly necessary to run bali-phy, but is necessary for building the
    tool <span class="application">draw-tree</span> that is used to
    draw consensus trees:

    </p><div class="itemizedlist"><ul class="itemizedlist" style="list-style-type: disc; "><li class="listitem"><p> The Cairo graphics library
	      (<a class="ulink" href="http://www.cairographics.org/" target="_top">Cairo</a>)
	  </p></li></ul></div><p>
	See also <a class="xref" href="#software_req" title="2.7. Additional required software">Section 2.7, &#8220;Additional required software&#8221;</a>.
    </p>
    </div>

	  
    </div>

    <div class="section"><div class="titlepage"><div><div><h3 class="title"><a name="idp57695152"></a>3.2. Quick Start</h3></div></div></div>
      
      <p>In order to compile the program on UNIX, first
	extract the source code archive, using a graphical archive
	manager, or the command-line tool <span class="command"><strong>tar</strong></span>:
	
	</p><pre class="screen"><code class="prompt">%</code> <strong class="userinput"><code>tar -zxf bali-phy-2.2.0.tgz</code></strong></pre><p>
	Then create a <span class="emphasis"><em>separate</em></span> build directory, enter it, and run the
	configure command:

	</p><pre class="screen"><code class="prompt">%</code> <strong class="userinput"><code>mkdir build</code></strong>
<code class="prompt">%</code> <strong class="userinput"><code>cd build</code></strong>
<code class="prompt">%</code> <strong class="userinput"><code>../bali-phy-2.2.0/configure --prefix=$HOME/Applications/bali-phy-2.2.2/ CXXFLAGS=-ftemplate-depth=256</code></strong></pre><p>
	
	If this command succeeds, then you can simply type

	</p><pre class="screen"><code class="prompt">%</code> <strong class="userinput"><code>make</code></strong>
<code class="prompt">%</code> <span class="command"><strong>make install</strong></span></pre><p>

	to build and install <span class="command"><strong>bali-phy</strong></span> and its
	associated tools and install it in <code class="filename">~/Applications/bali-phy-2.2.2</code>.
	To customize the compilation and installation process, read the
	following sections on supplying arguments to the
	<span class="command"><strong>configure</strong></span> script.
      </p>
    </div>

    <div class="section"><div class="titlepage"><div><div><h3 class="title"><a name="idp57703968"></a>3.3. Options to <span class="command"><strong>configure</strong></span></h3></div></div></div>
      

      <div class="section"><div class="titlepage"><div><div><h4 class="title"><a name="install-prefix"></a>3.3.1. Installing to another location</h4></div></div></div>
	
	<p>The configure script chooses to install
	  <span class="command"><strong>bali-phy</strong></span> in the directory
	  <code class="filename">/usr/local/</code> by default. You can install
	  executables to another directory <em class="replaceable"><code>dir</code></em>
	  by passing
	  <strong class="userinput"><code>--prefix=<em class="replaceable"><code>dir</code></em></code></strong>.
	  For example, in order to install
	  <span class="application">BAli-Phy</span> under
	  <code class="filename">~/local</code>, you can enter:

	  </p><pre class="screen"><code class="prompt">%</code> <strong class="userinput"><code>../bali-phy-2.2.0/configure --prefix=$HOME/local</code></strong></pre><p>
	</p>

      </div>

      <div class="section"><div class="titlepage"><div><div><h4 class="title"><a name="idp57710816"></a>3.3.2. Specifying where to find libraries and header files (e.g. Cairo)</h4></div></div></div>
	
	<p>You can instruct the compiler to look for include files
	  in directory <em class="replaceable"><code>dir</code></em> by passing
	  <em class="parameter"><code>--with-extra-includes=<em class="replaceable"><code>dir</code></em></code></em> to the <span class="command"><strong>configure</strong></span> script.
	</p>

	<p>You can instruct the compiler to look for libraries
	  files in directory
	  <code class="filename"><em class="replaceable"><code>dir</code></em></code> by
	  passing
	  <em class="parameter"><code>--with-extra-libs=<em class="replaceable"><code>dir</code></em></code></em> to the <span class="command"><strong>configure</strong></span> script.  
	</p> 

	<p>For example, if your system has Cairo installed in <code class="filename">/usr/local/</code>, then you might need to add "<strong class="userinput"><code>--with-extra-includes=/usr/local/include --with-extra-libs=/usr/local/lib</code></strong>" to the configure script arguments so that the compiler can find the Cairo include files and libraries.</p>
      </div>

      <div class="section"><div class="titlepage"><div><div><h4 class="title"><a name="idp57717040"></a>3.3.3. Selecting a non-default C++ compiler</h4></div></div></div>
	
	
	<p>The default C++ compiler is <span class="command"><strong>g++</strong></span>.  On
	  some systems, <span class="command"><strong>g++</strong></span> invokes GCC version 4.2,
	  and the correct compiler is called something else, such as
	  <span class="command"><strong>g++-4.8</strong></span>.  To use
	  <span class="command"><strong>g++-4.8</strong></span> as the C++ compiler when compiling
	  <span class="application">BAli-Phy</span>, you would set the CXX
	  environment variable as follows:
	  </p><pre class="screen"><code class="prompt">%</code> <strong class="userinput"><code>../bali-phy-2.2.0/configure CXX=g++-4.8</code></strong></pre><p>
	</p>
      </div>

      <div class="section"><div class="titlepage"><div><div><h4 class="title"><a name="idp57721568"></a>3.3.4. Passing flags to the C++ compiler</h4></div></div></div>
	
	
	<p>You can pass flags to the C++ compiler by setting the CXXFLAGS variable:
	  </p><pre class="screen"><code class="prompt">%</code> <strong class="userinput"><code>../bali-phy-2.2.0/configure CXXFLAGS="-march=native -g"</code></strong></pre><p>
	</p>
      </div>

      <div class="section"><div class="titlepage"><div><div><h4 class="title"><a name="idp57724080"></a>3.3.5. Example</h4></div></div></div>
	
	<p>
          All these options to <span class="command"><strong>configure</strong></span> can be
	  combined, as follows:
	  </p><pre class="screen"><code class="prompt">%</code> <strong class="userinput"><code>../bali-phy-2.2.0/configure --prefix=$HOME/local CXX=g++-4.8 CXXFLAGS="-march=native -g"</code></strong></pre><p>
	</p>
      </div>
    </div>
    <div class="section"><div class="titlepage"><div><div><h3 class="title"><a name="gitsource"></a>3.4. Generating the <span class="command"><strong>configure</strong></span> script
	and Makefiles (<span class="application">git</span> only)</h3></div></div></div>
       
      <p>Skip this step unless you are compiling a
	snapshot of the source code that you checked out using
	GIT. If you downloaded an official tar.gz archive of the
	source from the website, then it already includes these files.
      </p>

      <p>To generate these files, you need automake 1.8 (or higher)
	and autoconf 2.59 (or higher).  Run these commands in the top level
	directory of the repository that you checked out.</p><pre class="screen"><code class="prompt">%</code><strong class="userinput"><code> autoheader</code></strong>
<code class="prompt">%</code> <strong class="userinput"><code>aclocal -I m4</code></strong>
<code class="prompt">%</code> <strong class="userinput"><code>libtoolize</code></strong>
<code class="prompt">%</code> <strong class="userinput"><code>automake -a</code></strong>
<code class="prompt">%</code> <strong class="userinput"><code>autoconf</code></strong></pre><p>
        If your system has multiple versions of automake, then you may
	have to type e.g. <strong class="userinput"><code>automake-1.14 -a</code></strong> and
	<strong class="userinput"><code>aclocal-1.14</code></strong> instead in order to specify which
	version to use.
      </p>
    </div>



    <div class="section"><div class="titlepage"><div><div><h3 class="title"><a name="idp57735632"></a>3.5. Installing when compiling from source</h3></div></div></div>
      
      <p>
	After compiling <span class="application">BAli-Phy</span>, you can
	simply type <span class="command"><strong>make install</strong></span>.  This will copy the
	compiled executables to the installation directory (See <a class="xref" href="#install-prefix" title="3.3.1. Installing to another location">Section 3.3.1, &#8220;Installing to another location&#8221;</a>).
      </p>
    </div>

  </div>

  <div class="section"><div class="titlepage"><div><div><h2 class="title" style="clear: both"><a name="running"></a>4. Running the program</h2></div></div></div>
    

    <p>Here are some examples and explanations of how to run <span class="command"><strong>bali-phy</strong></span>.  You can get an overview of command line options (see <a class="xref" href="#command_line_options" title="4.6. Command-line options: An overview">Section 4.6, &#8220;Command-line options: An overview&#8221;</a>) by running <span class="command"><strong> bali-phy --help</strong></span>.</p>
<p>We recommend running multiple chains in parallel for each command, because 
</p><div class="orderedlist"><ol class="orderedlist" type="1"><li class="listitem">You can combine the samples, leading to faster run times.</li><li class="listitem">You can compare the runs to determine if the chains have converged.</li></ol></div><p>
</p>

    <div class="section"><div class="titlepage"><div><div><h3 class="title"><a name="idp57742800"></a>4.1. Quick Start</h3></div></div></div>
      
      <p>The simplest way to run <span class="command"><strong>BAli-Phy</strong></span> is
	to type all the arguments on the command line:

	</p><pre class="screen"><code class="prompt">%</code> <strong class="userinput"><code>bali-phy <em class="replaceable"><code>sequence-file</code></em></code></strong></pre><p>

	Here <em class="replaceable"><code>sequence-file</code></em> is a FastA or PHYLIP
	file containing the sequences you wish to analyze.  The filename should end
	in <strong class="userinput"><code>.fasta</code></strong> or <strong class="userinput"><code>.phy</code></strong> to
	indicate which format it is using.</p>

      <p>In this simple example, <span class="command"><strong>bali-phy</strong></span> automatically detects whether <em class="replaceable"><code>sequence-file</code></em> contains DNA, RNA, or Amino-Acids and uses default values for several command line options.  Thus, if <em class="replaceable"><code>sequence-file</code></em> contains DNA, then this is equivalent to the more verbose command line
	</p><pre class="screen"><code class="prompt">%</code> bali-phy <em class="replaceable"><code>sequence-file</code></em> --alphabet DNA --smodel TN --imodel RS07 --iterations=100000</pre><p>  Here the substitution model is Tamura-Nei, the insertion/deletion model is RS07, and the number of iterations is 100,000.  If <em class="replaceable"><code>sequence-file</code></em> contains amino acids, then the defaults will be:</p><pre class="screen"><code class="prompt">%</code> bali-phy <em class="replaceable"><code>sequence-file</code></em> --alphabet Amino-Acids --smodel LG --imodel RS07 --iterations=100000</pre><p> 
      </p>
    </div>

    <div class="section"><div class="titlepage"><div><div><h3 class="title"><a name="idp57751696"></a>4.2. Command line options</h3></div></div></div>
      

      <p>You can specify a more complex substitution model as follows (See <a class="xref" href="#subst_basic" title="7.2. Basic CTMC models">Section 7.2, &#8220;Basic CTMC models&#8221;</a>):

	</p><pre class="screen"><code class="prompt">%</code> <strong class="userinput"><code>bali-phy <em class="replaceable"><code>sequence-file</code></em> --smodel LG+gamma_inv</code></strong></pre><p> </p>
      <p>You may specify an indel model of <strong class="userinput"><code>none</code></strong>
	to fix the alignment to its initial value, and ignore information in shared insertions or deletions.
  	</p><pre class="screen"><code class="prompt">%</code> <strong class="userinput"><code>bali-phy <em class="replaceable"><code>sequence-file</code></em> --imodel none</code></strong></pre>
      <p>If you desire to use a codon model, you must specify the alphabet:
	</p><pre class="screen"><code class="prompt">%</code> <strong class="userinput"><code>bali-phy <em class="replaceable"><code>sequence-file</code></em> --smodel M0 --alphabet Codons</code></strong></pre><p>
      </p>
    </div>

    <div class="section"><div class="titlepage"><div><div><h3 class="title"><a name="idp57759248"></a>4.3. Multiple genes or data partitions</h3></div></div></div>
      
      <p>You may analyze multiple genes by putting each one it its own data partition:
	</p><pre class="screen"><code class="prompt">%</code> <strong class="userinput"><code>bali-phy <em class="replaceable"><code>sequence-file1</code></em> <em class="replaceable"><code>sequence-file2</code></em></code></strong></pre><p>
	You should put the data from the first gene in <em class="replaceable"><code>sequence-file1</code></em> and the second gene
	in <em class="replaceable"><code>sequence-file2</code></em>.  The sequence names in both files should be the same.  In this scenario, both genes share the same tree, but their alignments vary independently.  Furthermore, the branch lengths for each gene are scaled by an independent factor.  By default, each partition will have its own default alphabet, substitution model, insertion/deletion model, and tree length.    
      </p>

      <p>By default, each partition will recieve an independent copy of the model, and will not share parameter values:
</p><pre class="screen"><code class="prompt">%</code> <strong class="userinput"><code>bali-phy <em class="replaceable"><code>sequence-file1</code></em> <em class="replaceable"><code>sequence-file2</code></em> --smodel TN --imodel RS07</code></strong></pre><p>
However, you can select partition-specific values for 5 options: <strong class="userinput"><code>--smodel</code></strong>, <strong class="userinput"><code>--imodel</code></strong>, <strong class="userinput"><code>--alphabet</code></strong>, <strong class="userinput"><code>--same-scale</code></strong>, and <strong class="userinput"><code>--align-constraint</code></strong>.  For example, to specify different substitution models but the same alphabet:
</p><pre class="screen"><code class="prompt">%</code> <strong class="userinput"><code>bali-phy <em class="replaceable"><code>sequence-file1</code></em> <em class="replaceable"><code>sequence-file2</code></em> --smodel 1:TN --smodel 2:GTR --alphabet 1,2:DNA</code></strong></pre><p>
You can fix the alignment and ignore insertion/deletion information in one partition, while allowing the alignment to vary and using insertion/deletion information in another partition:
</p><pre class="screen"><code class="prompt">%</code> <strong class="userinput"><code>bali-phy <em class="replaceable"><code>sequence-file1</code></em> <em class="replaceable"><code>sequence-file2</code></em> --imodel 1:RS07 --imodel 2:none</code></strong></pre><p>
You can also specify that two partitions share a single copy of a single substitution model or indel model.  This reduces the number of parameters and also pools information between the partitions:
</p><pre class="screen"><code class="prompt">%</code> <strong class="userinput"><code>bali-phy <em class="replaceable"><code>sequence-file1</code></em> <em class="replaceable"><code>sequence-file2</code></em> --smodel 1,2:TN --imodel 1,2:RS07</code></strong></pre><p>
By default each partition has a separate scale, but you can force groups of partitions to share a scale. The name of the groups for the scale are not currently used, but may be used in later versions of the software:
</p><pre class="screen"><code class="prompt">%</code> <strong class="userinput"><code>bali-phy <em class="replaceable"><code>sequence-file1</code></em> <em class="replaceable"><code>sequence-file2</code></em> --smodel 1:TN --smodel 2:GTR --same-scale 1,2:groupname</code></strong></pre><p>
Finally, you may specify the option <strong class="userinput"><code>--traditional</code></strong>, or its short form <strong class="userinput"><code>-t</code></strong>.  This is the same as <strong class="userinput"><code>--imodel none</code></strong> and affects all partitions:
</p><pre class="screen"><code class="prompt">%</code> <strong class="userinput"><code>bali-phy <em class="replaceable"><code>sequence-file1</code></em> <em class="replaceable"><code>sequence-file2</code></em> --smodel 1:TN --smodel 2:GTR -t</code></strong></pre><p>
      </p>
    </div>

    <div class="section"><div class="titlepage"><div><div><h3 class="title"><a name="idp57778160"></a>4.4. Option files (Scripts)</h3></div></div></div>
      
      <p>
	In addition to using the command line, you may also specify
	options in a file. Using an option file can be more convenient
	if you are going to run the same analysis many times, or if
	the number of options is large. Furthermore, the option file
	may contain comments and blank lines.  Option files are a good
	to record what options you used in an analysis, and why.
      </p>

	<p>
	  An option file is specified with the command line option <strong class="userinput"><code>--config
	    <em class="replaceable"><code>file</code></em></code></strong> or <strong class="userinput"><code>-c<em class="replaceable"><code>file</code></em></code></strong>. If values
	    for an option are given both on the command line and
	    in an option file, then the command line value overrides
	    the value in the option file. 
	</p>
      <div class="section"><div class="titlepage"><div><div><h4 class="title"><a name="idp57781936"></a>4.4.1. Syntax</h4></div></div></div>
	

	<p>Option files use the same option names as the command
	line.  However, the syntax is different:  each option is given
	  on its own line using the syntax "<strong class="userinput"><code>option =
	    value</code></strong>" instead of the syntax "<strong class="userinput"><code>--option
	    value</code></strong>".  If the option has no value then it is
	  given using the syntax  "<strong class="userinput"><code>option =
	    option</code></strong>".  
	</p>
      </div>

      <div class="section"><div class="titlepage"><div><div><h4 class="title"><a name="idp57784960"></a>4.4.2. Example</h4></div></div></div>
	
	<p>
	  For example, consider the following
	  option file:
	  </p><pre class="programlisting">#select a data set to analyze
align = examples/EF-Tu/5d.fasta

#select an substitution model
smodel = log-normal_inv

#fix the alignment and do not model indels
traditional = traditional
	  </pre><p>
          The first option, <strong class="userinput"><code>align</code></strong>, is the name of
	  the sequence file, which has no name on the command line.
	  Lines that begin with # are comments, and blank lines are
	  ignored.  The option <strong class="userinput"><code>--traditional</code></strong> uses
	  the option name as the value, because it does not take a
	  value.  Thus, this configuration file corresponds to the
	  command line
	  </p><pre class="screen"><code class="prompt">%</code> <strong class="userinput"><code>bali-phy examples/EF-Tu/5d.fasta --smodel log-normal_inv --traditional</code></strong></pre><p>
	</p>
      </div>

      <div class="section"><div class="titlepage"><div><div><h4 class="title"><a name="idp57789072"></a>4.4.3. The configuration file</h4></div></div></div>
	
	<p>
	  The file <code class="filename">~/.bali-phy</code> is a special
	  option file called the <span class="emphasis"><em>configuration
	    file.</em></span>  If it exists, it is always loaded.
	  Options given on the command line or an option file 
	  override values given in <code class="filename">~/.bali-phy</code>. 
	</p>
      </div>
    </div>

  <div class="section"><div class="titlepage"><div><div><h3 class="title"><a name="examples"></a>4.5. Examples</h3></div></div></div>
    
    <p>Here are some examples which demonstrate how to run
      <span class="application">BAli-Phy</span>.  In order to run these
      examples, you must find the <code class="filename">examples/</code>
      directory which contains the example files.  If you downloaded
      executables and extracted them in the
      <code class="filename">~/Applications</code> directory, then the
      <code class="filename">examples/</code> directory will be found at 
      <code class="filename">~/Applications/bali-phy-2.2.2/share/bali-phy/examples/</code>.
    </p>

    <p>Also note that <span class="command"><strong>bali-phy</strong></span>
      does <span class="emphasis"><em>not</em></span> run until it
      is "finished", but continues to gather samples until the user
      determines that enough samples have been gathered, and stops it.
      Thus, it is useful to continually examine the output files while
      the program is running. 
    </p>

    <div class="example"><a name="idp57797120"></a><p class="title"><b>Example 1. No frills</b></p><div class="example-contents">
      
      <p>Here we analyze the EF-Tu 5-taxon data set provided with
	the software.</p>
      <pre class="screen"> <code class="prompt">%</code> <strong class="userinput"><code>bali-phy ~/Applications/bali-phy-2.2.2/share/bali-phy/examples/EF-Tu/5d.fasta</code></strong></pre>
    </div></div><br class="example-break">

    <div class="example"><a name="idp57800432"></a><p class="title"><b>Example 2. Multiple-Rate Substitution Model</b></p><div class="example-contents">
      
      <p>We now modify the previous example by changing the
	substitution model to allow log-normal-distributed rate variation and
	invariant sites.  The amount of rate variation and the fraction of
	invariant sites are estimated</p>
      
      <pre class="screen"> <code class="prompt">%</code> <strong class="userinput"><code>bali-phy ~/Applications/bali-phy-2.2.2/share/bali-phy/examples/EF-Tu/5d.fasta --smodel log-normal_inv --randomize-alignment</code></strong></pre></div></div><br class="example-break">

    <div class="example"><a name="idp57803408"></a><p class="title"><b>Example 3. Fixed alignment</b></p><div class="example-contents">
      
      <p>Here we use the 5S rRNA 5-taxon data set provided with
	the software.  The alignment is fixed and the <strong class="userinput"><code>--traditional</code></strong>
	option is used, making indels non-informative.  In addition,
	nucleotide frequencies are fixed to empirical values estimated
	from the input sequences. 
      </p> 

      <pre class="screen"> <code class="prompt">%</code> <strong class="userinput"><code>bali-phy ~/Applications/bali-phy-2.2.2/share/bali-phy/examples/5S-rRNA/5d.fasta --smodel F=constant --traditional</code></strong></pre>

    </div></div><br class="example-break">
  </div>

  <div class="section"><div class="titlepage"><div><div><h3 class="title"><a name="command_line_options"></a>4.6. Command-line options: An overview</h3></div></div></div>
    

    <p>You can get an up-to-date overview of these command line options by running <span class="command"><strong> bali-phy --help</strong></span>.</p>
    <div class="section"><div class="titlepage"><div><div><h4 class="title"><a name="idp57808176"></a>4.6.1. General options</h4></div></div></div>
      
      
      <div class="variablelist"><table border="0" class="variablelist"><colgroup><col align="left" valign="top"><col></colgroup><tbody><tr><td><p><span class="term"><code class="option">-h, --help</code></span></p></td><td><p>Show help message.</p></td></tr><tr><td><p><span class="term"><code class="option">-v, --version</code></span></p></td><td><p>Show version information.</p></td></tr><tr><td><p><span class="term"><code class="option">-c <em class="replaceable"><code>file</code></em>, --config <em class="replaceable"><code>file</code></em></code></span></p></td><td><p>Option file to read.</p></td></tr><tr><td><p><span class="term"><code class="option">--show-only</code></span></p></td><td><p>Analyze initial values and exit.</p></td></tr><tr><td><p><span class="term"><code class="option">--seed <em class="replaceable"><code>seed</code></em></code></span></p></td><td><p>Use the specified seed to initialize the
	      random number generator.</p></td></tr><tr><td><p><span class="term"><code class="option">--name <em class="replaceable"><code>string</code></em></code></span></p></td><td><p>Specify the name for the analysis directory.</p></td></tr><tr><td><p><span class="term"><code class="option">-t, --traditional</code></span></p></td><td><p>Fix the alignment and don't model indels.</p></td></tr></tbody></table></div>
    </div>

    <div class="section"><div class="titlepage"><div><div><h4 class="title"><a name="idp57821040"></a>4.6.2. MCMC options</h4></div></div></div>
      

      <div class="variablelist"><table border="0" class="variablelist"><colgroup><col align="left" valign="top"><col></colgroup><tbody><tr><td><p><span class="term"><code class="option">-i, --iterations <em class="replaceable"><code>number=100000</code></em></code></span></p></td><td><p>Specify the number of iterations to run.</p></td></tr><tr><td><p><span class="term"><code class="option">--pre-burnin <em class="replaceable"><code>iterations=3</code></em></code></span></p></td><td><p>Iterations to refine initial tree.</p></td></tr><tr><td><p><span class="term"><code class="option">--subsample <em class="replaceable"><code>factor=1</code></em></code></span></p></td><td><p>Specify a factor by which to subsample.</p></td></tr><tr><td><p><span class="term"><code class="option">--enable <em class="replaceable"><code>move</code></em></code></span></p></td><td>
	    <p>Enable a comma-separated list of transition kernels.</p>
	  </td></tr><tr><td><p><span class="term"><code class="option">--disable <em class="replaceable"><code>move</code></em></code></span></p></td><td>
	    <p>Disable a comma-separated list of transition kernels.</p>
	</td></tr></tbody></table></div>
    </div>


    <div class="section"><div class="titlepage"><div><div><h4 class="title"><a name="idp57801152"></a>4.6.3. Parameter options</h4></div></div></div>
      
      <div class="variablelist"><table border="0" class="variablelist"><colgroup><col align="left" valign="top"><col></colgroup><tbody><tr><td><p><span class="term"><code class="option">--randomize-alignment</code></span></p></td><td><p>Randomly re-align sequences before use.</p></td></tr><tr><td><p><span class="term"><code class="option">--tree <em class="replaceable"><code>file</code></em></code></span></p></td><td><p>Specify file with initial tree.</p></td></tr><tr><td><p><span class="term"><code class="option">--set <em class="replaceable"><code>key=value</code></em></code></span></p></td><td><p>Set <em class="replaceable"><code>key</code></em>=<em class="replaceable"><code>value</code></em>.</p></td></tr><tr><td><p><span class="term"><code class="option">--initial-value <em class="replaceable"><code>parameter[=value]</code></em></code></span></p></td><td><p>Set <em class="replaceable"><code>parameter</code></em>=<em class="replaceable"><code>initial value</code></em>.</p></td></tr><tr><td><p><span class="term"><code class="option">--frequencies <em class="replaceable"><code>frequencies</code></em></code></span></p></td><td><p>Specify initial frequencies: 'uniform','nucleotides', or a comma-separated list of frequencies.</p></td></tr></tbody></table></div>

    </div>

    <div class="section"><div class="titlepage"><div><div><h4 class="title"><a name="idp57842352"></a>4.6.4. Model options</h4></div></div></div>
      

      <div class="variablelist"><table border="0" class="variablelist"><colgroup><col align="left" valign="top"><col></colgroup><tbody><tr><td><p><span class="term"><code class="option">--alphabet <em class="replaceable"><code>name</code></em></code></span></p></td><td><p>Specify the alphabet: DNA, RNA, Amino-Acids, Amino-Acids+stop, Triplets, Codons, or Codons + stop.</p></td></tr><tr><td><p><span class="term"><code class="option">--smodel <em class="replaceable"><code>name</code></em></code></span></p></td><td><p>Specify the substitution model.</p></td></tr><tr><td><p><span class="term"><code class="option">--imodel <em class="replaceable"><code>name</code></em></code></span></p></td><td><p>Specify the indel model.</p></td></tr><tr><td><p><span class="term"><code class="option">--branch-prior <em class="replaceable"><code>name</code></em></code></span></p></td><td><p>Exponential or Gamma.</p></td></tr><tr><td><p><span class="term"><code class="option">--same-scale <em class="replaceable"><code>specification</code></em></code></span></p></td><td><p>Which partitions have the same scale?</p></td></tr><tr><td><p><span class="term"><code class="option">--align-constraint <em class="replaceable"><code>filename</code></em></code></span></p></td><td><p>File with alignment constraints.</p></td></tr></tbody></table></div>

    </div>

    </div>
  </div>

  <div class="section"><div class="titlepage"><div><div><h2 class="title" style="clear: both"><a name="input"></a>5. Input</h2></div></div></div>
    

    <div class="section"><div class="titlepage"><div><div><h3 class="title"><a name="idp57854288"></a>5.1. Sequence formats</h3></div></div></div>
      
      <p><span class="application">BAli-Phy</span> can read in sequences
	and alignments in both FastA and PHYLIP formats.  Filenames for
	FastA files should end in <strong class="userinput"><code>.fasta</code></strong>,
	<strong class="userinput"><code>.mpfa</code></strong>, <strong class="userinput"><code>.fna</code></strong>,
	<strong class="userinput"><code>.fas</code></strong>, <strong class="userinput"><code>.fsa</code></strong>, or
	<strong class="userinput"><code>.fa</code></strong>.  Filenames for PHYLIP files should
	end in <strong class="userinput"><code>.phy</code></strong>.  If one of these extensions
	is not used, then <span class="application">BAli-Phy</span> will
	attempt to guess which format is being used.
      </p>

    </div>

    <div class="section"><div class="titlepage"><div><div><h3 class="title"><a name="idp57860064"></a>5.2. Is my data set too large?</h3></div></div></div>
      

      <p>Large data sets run more slowly than small data
	sets. We recommend a conservative starting point with few taxa
	and short sequence lengths.  You can then increase the size of
	your data set until a balance between speed and size is
	reached.</p> 

      <p>The number of samples that you need depends on whether you
	are primarily interested in obtaining a point estimate or in
	obtaining detailed measures of confidence and uncertainty.  For
	detailed measures of confidence and uncertainty you should
	obtain a minimum of 10,000 samples after the Markov chain
	converges.  For an estimate, you don't need very many samples
	after convergence.  (But you may need many samples to be sure
	that you've converged!)
      </p>

      <div class="note" style="margin-left: 0.5in; margin-right: 0.5in;"><h3 class="title">Computing clusters can speed up
      MCMC analysis</h3>
      <p>
	Running <span class="command"><strong>bali-phy</strong></span> on a computing cluster is
	not necessary, but can speed up the analysis dramatically.
	This is because a cluster allows you to run several
	<span class="emphasis"><em>independent</em></span> MCMC chains in parallel and pool the resulting
      samples.</p> 

	<p>This approach to parallel
	computation is sometimes more efficient than MCMCMC-based
	parallelism involving heated chains.  It is equivalent to
	running MCMCMC with no temperature difference between
	chains, with the exception that it allows results from
	<span class="emphasis"><em>all</em></span> chains to be used, instead of just
	results from the single "cold" chain.  Thus, if you run 10
	independent chains in parallel, then you may gather
	samples 10 times faster that a single chain.
      </p>
      </div>

      <div class="section"><div class="titlepage"><div><div><h4 class="title"><a name="idp57866416"></a>5.2.1. Too many taxa?</h4></div></div></div>
	

	<p><span class="application">BAli-Phy</span> is quite CPU
	  intensive, and so we recommend using 50 or fewer taxa in order
	  to limit the time required to accumulate enough MCMC
	  samples. (Despite this recommendation, data sets with more 
	  than 100 taxa have occasionally been known to converge.)
	  We recommend initially pruning as many taxa as possible
	  from your data set, then adding some back if the MCMC is not
	  too slow.</p> 

      </div>

      <div class="section"><div class="titlepage"><div><div><h4 class="title"><a name="idp57868608"></a>5.2.2. Sequences too long?</h4></div></div></div>
	

	<p>Aligning just a pair of sequences takes <mml:math xmlns:mml="http://www.w3.org/1998/Math/MathML"><mml:mi>O</mml:mi><mml:mo>(</mml:mo><mml:msup><mml:mi>L</mml:mi> <mml:mn>2 </mml:mn></mml:msup><mml:mo>)</mml:mo></mml:math> time
	  and memory, where <mml:math xmlns:mml="http://www.w3.org/1998/Math/MathML"><mml:mi>L</mml:mi></mml:math> represents the sequence length.  Therefore
	  sequences longer than (say) 1000 letters become increasingly
	  impractical.  However, you might try to see how long you can make your
	  sequences before you run out of memory, or the program
	  becomes too slow.</p>

	<p>For multi-gene analyses, two separate data partitions
	  (i.e. genes) of 500 letters will be twice as fast
	  to align as one data partition of 1000 letters. So, it may be possible
	  to analyze several genes as long as each gene individually
	  is not too long.</p> 

	<p>You can speed up alignment for long genes by specifying
	  alignment constraints (See <a class="xref" href="#alignment_constraints" title="9. Alignment constraints">Section 9, &#8220;Alignment constraints&#8221;</a>).
	  Ideally, 10 evenly spaced constraints should reduce the cost of
	  re-aligning a sequence by a factor of 10.
	</p>

	<p>Also, note that you can sometimes speed up the analysis
	  of protein sequences by coding them as amino acids or codons, rather 
	  than nucleotides. This is because it decreases the sequence
	  length.
	</p> 
      </div>

    </div>

  </div>

  <div class="section"><div class="titlepage"><div><div><h2 class="title" style="clear: both"><a name="output"></a>6. Output</h2></div></div></div>
    

    <div class="section"><div class="titlepage"><div><div><h3 class="title"><a name="idp57876000"></a>6.1. Output directory</h3></div></div></div>
      
      <p><span class="application">BAli-Phy</span> creates a new
	directory to store its output files each time it is run.  By default, the
	directory name is the name of the sequence file, with a number
	added on the end to make it unique. <span class="application">BAli-Phy</span>
	first checks  if there is already a directory called
	<code class="filename"><em class="replaceable"><code>file</code></em>-1/</code>, and then moves on to
	<code class="filename"><em class="replaceable"><code>file</code></em>-2/</code>, etc. until it finds an
	unused directory name.</p> 
      
      <p>You can specify a different name to use instead of the
	sequence-file name by using the <strong class="userinput"><code>--name</code></strong> option.</p>
    </div>

    <div class="section"><div class="titlepage"><div><div><h3 class="title"><a name="idp57881168"></a>6.2. Output options</h3></div></div></div>

    <p>In order to infer ancestral sequences at internal nodes, add the option <strong class="userinput"><code>--set log-ancestral=1</code></strong>.  This will lead to the creation of an additional output file called <code class="filename">C1.P1.ancestral.fastas</code>. This file contains a sampled alignment for each iteration, where N's are replaced with a letter randomly sampled from the posterior distribution.</p>
    </div>

    <div class="section"><div class="titlepage"><div><div><h3 class="title"><a name="idp57883616"></a>6.3. Output files</h3></div></div></div>
      
      <p><span class="application">BAli-Phy</span> writes the following output
	files inside the directory that it creates:</p>
      
      <div class="variablelist"><table border="0" class="variablelist"><colgroup><col align="left" valign="top"><col></colgroup><tbody><tr><td><p><span class="term">C1.out</span></p></td><td>
	    <p>Iteration numbers, probabilities, success probabilities for transition kernels, etc..</p>
	  </td></tr><tr><td><p><span class="term">C1.P<em class="replaceable"><code>p</code></em>.fastas</span></p></td><td>
	    <p>Sampled alignments for partition <em class="replaceable"><code>p</code></em></p>
	  </td></tr><tr><td><p><span class="term">C1.err</span></p></td><td>
	    <p>Log file for hopefully irrelevant error messages.</p>
	  </td></tr><tr><td><p><span class="term">C1.MAP</span></p></td><td>
	    <p>Successive estimates of the MAP point.</p>
	  </td></tr><tr><td><p><span class="term">C1.p</span></p></td><td>
	    <p>Scalar parameters: indel and substitution parameters, etc.</p>
	  </td></tr><tr><td><p><span class="term">C1.trees</span></p></td><td>
	    <p>Tree samples: one sample per line, in Newick format.</p>
	  </td></tr></tbody></table></div>
      
      <p>For the last two files, each line in these files corresponds to one iteration.</p>

      <div class="section"><div class="titlepage"><div><div><h4 class="title"><a name="idp57896288"></a>6.3.1. Field names in <code class="filename">C1.p</code></h4></div></div></div>

      <p>This section explains the meaning of the various field names in the file <code class="filename">C1.p</code>.</p>


      <div class="section"><div class="titlepage"><div><div><h5 class="title"><a name="idp57897920"></a>6.3.1.1. Computed parameter names</h5></div></div></div>
      <div class="variablelist"><table border="0" class="variablelist"><colgroup><col align="left" valign="top"><col></colgroup><tbody><tr><td><p><span class="term">prior</span></p></td><td>
	    <p>The log prior probability.  This includes the probability of the alignment, since the alignment is not observed.</p>
	  </td></tr><tr><td><p><span class="term">prior_A<em class="replaceable"><code>n</code></em></span></p></td><td>
	    <p>The log of the probability <mml:math xmlns:mml="http://www.w3.org/1998/Math/MathML"><mml:mi>Pr</mml:mi><mml:mo>(</mml:mo><mml:msub><mml:mi>A</mml:mi> <mml:mi>n</mml:mi></mml:msub><mml:mo lspace="0em" rspace="0em">&#8739;</mml:mo><mml:mi>&#964;</mml:mi><mml:mo>,</mml:mo><mml:mi>T</mml:mi><mml:mo>,</mml:mo><mml:mi>&#923;</mml:mi><mml:mo>)</mml:mo></mml:math> of the alignment <mml:math xmlns:mml="http://www.w3.org/1998/Math/MathML"><mml:msub><mml:mi>A</mml:mi> <mml:mi>n</mml:mi></mml:msub></mml:math> of the <mml:math xmlns:mml="http://www.w3.org/1998/Math/MathML"><mml:mi>n</mml:mi></mml:math>th partition, given the topology <mml:math xmlns:mml="http://www.w3.org/1998/Math/MathML"><mml:mi>&#964;</mml:mi></mml:math>, the branch lengths <mml:math xmlns:mml="http://www.w3.org/1998/Math/MathML"><mml:mi>T</mml:mi></mml:math>, and insertion-deletion process parameters <mml:math xmlns:mml="http://www.w3.org/1998/Math/MathML"><mml:mi>&#923;</mml:mi></mml:math>.  This log probability is the probabilistic equivalent of a gap penalty on the alignment <mml:math xmlns:mml="http://www.w3.org/1998/Math/MathML"><mml:msub><mml:mi>A</mml:mi> <mml:mi>n</mml:mi></mml:msub></mml:math> given the scoring parameters <mml:math xmlns:mml="http://www.w3.org/1998/Math/MathML"><mml:mi>&#923;</mml:mi></mml:math>.</p>
	  </td></tr><tr><td><p><span class="term">likelihood</span></p></td><td>
	    <p>The log of the likelihood.  Conditional on the alignment, this is determined entirely by the substitution model, and ignores insertions and deletions.  This is the probabilistic equivalent of the mismatch penalty.</p>
	  </td></tr><tr><td><p><span class="term">logp</span></p></td><td>
	    <p>The log of the probability.  The probability is the product of the prior and the likelihood.</p>
	  </td></tr><tr><td><p><span class="term">|A|</span></p></td><td>
	    <p>The total number of alignment columns across all partitions.</p>
	  </td></tr><tr><td><p><span class="term">#indels<em class="replaceable"><code>n</code></em></span></p></td><td>
	    <p>The number of indel events in partition <em class="replaceable"><code>n</code></em>, if we group adjacent indels that occur on the same branch.</p>
	  </td></tr><tr><td><p><span class="term">#indels</span></p></td><td>
	    <p>The total number of indel events across all partitions, if we group adjacent indels that occur on the same branch.</p>
	  </td></tr><tr><td><p><span class="term">|indels<em class="replaceable"><code>n</code></em>|</span></p></td><td>
	    <p>The length of indel events in partition <em class="replaceable"><code>n</code></em>, if we group adjacent indels that occur on the same branch.</p>
	  </td></tr><tr><td><p><span class="term">|indels|</span></p></td><td>
	    <p>The total length of indel events across all partitions, if we group adjacent indels that occur on the same branch.</p>
	  </td></tr><tr><td><p><span class="term">#substs<em class="replaceable"><code>n</code></em></span></p></td><td>
	    <p>The unweighted parsimony score for substitutions in partition <em class="replaceable"><code>n</code></em>.</p>
	  </td></tr><tr><td><p><span class="term">#substs</span></p></td><td>
	    <p>The total unweighted parsimony score for substitutions across all partitions.</p>
	  </td></tr><tr><td><p><span class="term">|T|</span></p></td><td>
	    <p>For a single-partition analysis, the sum of branch lengths.  For a multi-partition analysis, a weighted average of this sum across partitions.</p>
	  </td></tr></tbody></table></div>
      </div>

      <div class="section"><div class="titlepage"><div><div><h5 class="title"><a name="idp57930256"></a>6.3.1.2. Model parameter names</h5></div></div></div>

      <p>The prefixes "S<em class="replaceable"><code>n</code></em>::" and "I<em class="replaceable"><code>n</code></em>::" will be dropped if not necessary to disambiguate parameters with the same name in different sub-models.</p>

      <div class="variablelist"><table border="0" class="variablelist"><colgroup><col align="left" valign="top"><col></colgroup><tbody><tr><td><p><span class="term">mu<em class="replaceable"><code>n</code></em></span></p></td><td>
	    <p>The average number of substitutions per branch.  The <em class="replaceable"><code>n</code></em>th scale parameter ordinarily applies to the <em class="replaceable"><code>n</code></em>th partition, unless multiple partitions are forced to have the same branch lengths using <strong class="userinput"><code>--same-scale</code></strong>.</p>
	  </td></tr><tr><td><p><span class="term">S<em class="replaceable"><code>n</code></em>::<em class="replaceable"><code>name</code></em></span></p></td><td>
	    <p>Parameter <em class="replaceable"><code>name</code></em> in the <em class="replaceable"><code>n</code></em>th substitution model.</p>
	  </td></tr><tr><td><p><span class="term">I<em class="replaceable"><code>n</code></em>::<em class="replaceable"><code>name</code></em></span></p></td><td>
	    <p>Parameter <em class="replaceable"><code>name</code></em> in the <em class="replaceable"><code>n</code></em>th insertion/deletion model.</p>
	  </td></tr></tbody></table></div>
      </div>
      </div>
    </div>

    <div class="section"><div class="titlepage"><div><div><h3 class="title"><a name="idp57943552"></a>6.4. Summarizing the output</h3></div></div></div>
      

      <p>This section is primarily oriented to extracting estimates from output files.  See <a class="xref" href="#mixing_and_convergence" title="10. Convergence and Mixing: Is it done yet?">Section 10, &#8220;Convergence and Mixing: Is it done yet?&#8221;</a> for methods of determine effective sample sizes, and for checking mixing and convergence.</p>

      <div class="section"><div class="titlepage"><div><div><h4 class="title"><a name="idp57945936"></a>6.4.1. Finding the consensus tree (<code class="filename">C1.trees</code>)</h4></div></div></div>
	
	<p>
To compute the majority consensus tree, do the following.  (The
program <a class="ulink" href="http://tree.bio.ed.ac.uk/software/figtree/" target="_top">FigTree</a>
allows you to view the resulting tree file graphically.)
</p><pre class="screen"><code class="prompt">%</code> trees-consensus C1.trees &gt; <code class="filename">c50.PP.tree</code></pre><p>
You can (and should) pool results from different MCMC runs by adding
multiple tree sample files on the command line.  The different MCMC
runs should have the same input files and parameters.
</p><pre class="screen"><code class="prompt">%</code> trees-consensus <em class="replaceable"><code>dir1</code></em>/C1.trees <em class="replaceable"><code>dir2</code></em>/C1.trees &gt; <code class="filename">c50.PP.tree</code></pre><p>
By default, the first 10% of tree samples are skipped as burn-in.  You
can specify the number of samples (e.g. 1000) to skip by adding the
options <strong class="userinput"><code>-s1000</code></strong> or
<strong class="userinput"><code>--skip=1000</code></strong>.  You may also specify a percentage
of all samples:
</p><pre class="screen"><code class="prompt">%</code> trees-consensus -s20% C1.trees &gt; <code class="filename">c50.PP.tree</code></pre><p>
To discard some samples, keeping (say) every 10th sample, you may add
the options <strong class="userinput"><code>-x10</code></strong> or
<strong class="userinput"><code>--sub-sample=10</code></strong>.  This can make the program a
lot faster, at the possible expense of some loss in accuracy.
</p><pre class="screen"><code class="prompt">%</code> trees-consensus -s20% -x10 C1.trees &gt; <code class="filename">c50.PP.tree</code></pre><p>
By default, splits are included in the consensus tree if they have a
PP greater than 0.5.  You can specify a more stringent level
(e.g. 0.66) by adding the option
<strong class="userinput"><code>--consensus-PP=0.66</code></strong> as follows:
</p><pre class="screen"><code class="prompt">%</code> trees-consensus -s20% -x10 --consensus-PP=0.66 C1.trees &gt; <code class="filename">c66.PP.tree</code></pre><p>
You may also make the program write directly to the output file
(e.g. <code class="filename">c66.PP.tree</code>) by using the more general form
<strong class="userinput"><code>--consensus-PP=0.66:c66.PP.tree</code></strong>.  Leaving off 
the "<strong class="userinput"><code>:c66.PP.tree</code></strong>" part (as we did above) or specifying
"<strong class="userinput"><code>:-</code></strong>" sends the output to the standard output
(e.g. the terminal, if not redirected). 
</p><pre class="screen"><code class="prompt">%</code> trees-consensus -s20% -x10 C1.trees --consensus-PP=0.66:<code class="filename">c66.PP.tree</code></pre><p>
You can supply multiple levels and filenames separated by commas.
This is faster than running the program multiple times with different
consensus levels.
</p><pre class="screen"><code class="prompt">%</code> trees-consensus -s20% -x10 C1.trees --consensus-PP=0.5:<code class="filename">c50.PP.tree</code>,0.66:<code class="filename">c66.PP.tree</code></pre><p>
Finally, you may use the option <strong class="userinput"><code>--consensus=</code></strong>
instead of the option <strong class="userinput"><code>--consensus-PP=</code></strong> if you do
not wish the resulting tree to contain embedded posterior
probabilities on branches, as well as branch lengths.
</p><pre class="screen"><code class="prompt">%</code> trees-consensus -s20% -x10 C1.trees --consensus=0.5:<code class="filename">c50.PP.tree</code>,0.66:<code class="filename">c66.PP.tree</code></pre><p>
Both the <strong class="userinput"><code>--consensus=</code></strong> and 
<strong class="userinput"><code>--consensus-PP=</code></strong> options may be given simultaneously.
</p>

<p>
  See <strong class="userinput"><code>trees-consensus --help</code></strong> for a complete list of options.
</p>
      </div>

      <div class="section"><div class="titlepage"><div><div><h4 class="title"><a name="idp57966720"></a>6.4.2. Finding the M.A.P. topology (<code class="filename">C1.trees</code>)</h4></div></div></div>
	
	<p>
To compute the <span class="emphasis"><em>maximum a posteriori</em></span> tree topology do:
</p><pre class="screen"><code class="prompt">%</code> trees-consensus --skip=<em class="replaceable"><code>burnin</code></em> C1.trees --map-tree=<code class="filename">MAP.tree</code></pre><p>	
The MAP topology may be used instead of a consensus tree when a fully resolved (e.g. bifurcating) tree is required.  However, when the topology has many tips, each topology may be sampled only once, leading to low quality estimates of the MAP topology.
</p>
<p>The program <a class="ulink" href="http://tree.bio.ed.ac.uk/software/figtree/" target="_top">FigTree</a> allows you to view the consensus tree graphically.
</p>
      </div>

      <div class="section"><div class="titlepage"><div><div><h4 class="title"><a name="idp57972656"></a>6.4.3. Checking topology convergence (<code class="filename">C1.trees</code>)</h4></div></div></div>
	
<p>
</p><pre class="screen"><code class="prompt">%</code> trees-bootstrap <em class="replaceable"><code>dir1</code></em>/C1.trees <em class="replaceable"><code>dir2</code></em>/C1.trees</pre><p>	
This command computes the effective sample size for the posterior probability of each split.  It also computes the Average Standard Deviation of Split Frequencies (ASDSF) between two or more independent runs.</p>

<p>See <a class="xref" href="#mixing_and_convergence" title="10. Convergence and Mixing: Is it done yet?">Section 10, &#8220;Convergence and Mixing: Is it done yet?&#8221;</a> for more information.
</p>  
      </div>

      <div class="section"><div class="titlepage"><div><div><h4 class="title"><a name="idp57978384"></a>6.4.4. Summarizing numerical parameters (<code class="filename">C1.p</code>)</h4></div></div></div>
	
<p>
This command gives a median and confidence interval, ESS, and a stabilization time:
</p><pre class="screen"><code class="prompt">%</code> statreport C1.p &gt; Report </pre><p>	
This command compares multiple runs to give PSRF and joint ESS values as well:
</p><pre class="screen"><code class="prompt">%</code> statreport C1.p C2.p &gt; Report </pre><p>	
The program <a class="ulink" href="http://tree.bio.ed.ac.uk/software/tracer/" target="_top">Tracer</a> allows you to view the same summaries graphically.</p>

<p>See <a class="xref" href="#mixing_and_convergence" title="10. Convergence and Mixing: Is it done yet?">Section 10, &#8220;Convergence and Mixing: Is it done yet?&#8221;</a> for more information.
</p>
      </div>

      <div class="section"><div class="titlepage"><div><div><h4 class="title"><a name="idp57984352"></a>6.4.5. Computing an alignment using Posterior Decoding (<code class="filename">C1.P<em class="replaceable"><code>p</code></em>.fastas</code>)</h4></div></div></div>
	
<p>
</p><pre class="screen"><code class="prompt">%</code> cut-range --skip=<em class="replaceable"><code>burn-in</code></em> &lt; C1.P<em class="replaceable"><code>p</code></em>.fastas | alignment-max &gt; P<em class="replaceable"><code>p</code></em>-max.fasta</pre><p>
You can use the program <a class="ulink" href="http://pbil.univ-lyon1.fr/software/seaview.html" target="_top">seaview</a> to view the alignment graphically.
</p>
      </div>      

      <div class="section"><div class="titlepage"><div><div><h4 class="title"><a name="idp57990416"></a>6.4.6. Find the alignment from the maximum a posterior (MAP) point (<code class="filename">C1.MAP</code>)</h4></div></div></div>
	
<pre class="screen"><code class="prompt">%</code> alignment-find &lt; C1.MAP &gt; P1-MAP.fasta</pre>

      </div>      

      <div class="section"><div class="titlepage"><div><div><h4 class="title"><a name="idp57993664"></a>6.4.7. Create an Au (Alignment Uncertainty) plot (<code class="filename">C1.P<em class="replaceable"><code>p</code></em>.fastas</code>)</h4></div></div></div>
	
<p>To annotate a specific alignment <em class="replaceable"><code>alignment</code></em>.fasta, choose a fully resolved tree estimate <em class="replaceable"><code>tree</code></em>:
</p><pre class="screen"><code class="prompt">%</code> cut-range --skip=<em class="replaceable"><code>burn-in</code></em> &lt; C1.P<em class="replaceable"><code>p</code></em>.fastas | alignment-gild <em class="replaceable"><code>alignment</code></em>.fasta <em class="replaceable"><code>tree</code></em>  &gt; <em class="replaceable"><code>alignment</code></em>-AU.prob
<code class="prompt">%</code> alignment-draw <em class="replaceable"><code>alignment</code></em>.fasta --AU <em class="replaceable"><code>alignment</code></em>-AU.prob &gt; <em class="replaceable"><code>alignment</code></em>-AU.html</pre><p>
The majority consensus tree is usually not fully resolved, so we recommend using the MAP topology instead.
</p>
      </div>


    </div>


  <div class="section"><div class="titlepage"><div><div><h3 class="title"><a name="analysis"></a>6.5. Summarizing the output - scripted</h3></div></div></div>
    
    
    <p>
      Instead of manually running each of the steps to analyze the
      output files, you may instead run the PERL script
      <span class="command"><strong>bp-analyze.pl</strong></span> to execute these commands.  You
      may run it inside the output directory, like this:
</p><pre class="screen"><code class="prompt">%</code> bp-analyze.pl --burnin=<em class="replaceable"><code>iterations</code></em></pre><p>
      The script will create an HTML
      page <code class="filename">Results/index.html</code> that summarizes the
      posterior distribution.
    </p>
    <p>
      You may also run it with one or more output directories as
      arguments, like this:
</p><pre class="screen"><code class="prompt">%</code> bp-analyze.pl --burnin=<em class="replaceable"><code>iterations</code></em> <em class="replaceable"><code>directory</code></em>-1/ <em class="replaceable"><code>directory</code></em>-2/</pre><p>
      In this case, output from multiple runs will be used to assess convergence and mixing, as well as to increase the precision of the estimates.
    </p>

    <div class="section"><div class="titlepage"><div><div><h4 class="title"><a name="idp58009536"></a>6.5.1. Meaning of generated files</h4></div></div></div>
      
    <p>The <code class="filename">Results/</code> directory will contain
    the following useful files:</p>

      <div class="variablelist"><table border="0" class="variablelist"><colgroup><col align="left" valign="top"><col></colgroup><tbody><tr><td><p><span class="term">Report</span></p></td><td>
	    <p>A summary of numerical parameters: credible
	    intervals and mixing.</p>
	</td></tr><tr><td><p><span class="term">consensus</span></p></td><td>
	    <p>A summary of supported splits (clades). </p>
	</td></tr><tr><td><p><span class="term">c-levels.plot</span></p></td><td>
	    <p>The number of splits (clades) supported at each LOD level.</p>
	</td></tr><tr><td><p><span class="term">c50.tree</span></p></td><td>	<p>The majority consensus topology + branch lengths (Newick format)</p> 
	</td></tr><tr><td><p><span class="term">c50.PP.tree</span></p></td><td>
	<p>The majority consensus topology + branch lengths +
	Posterior Probabilities (Newick format)</p> 
	</td></tr><tr><td><p><span class="term">MAP.tree</span></p></td><td>
	    <p>An estimate of the MAP topology + branch lengths (Newick format)</p>
	</td></tr></tbody></table></div>
      <p> 
	The following files will be generated to summarize alignment uncertainty, unless the analysis uses a fixed alignment.

      </p>

      <div class="variablelist"><table border="0" class="variablelist"><colgroup><col align="left" valign="top"><col></colgroup><tbody><tr><td><p><span class="term">MAP.fasta</span></p></td><td>
	    <p>An estimate of the MAP alignment.
	</p></td></tr><tr><td><p><span class="term">P<em class="replaceable"><code>p</code></em>-max.fasta</span></p></td><td>
	    <p>An estimate of the alignment for partition
	    <em class="replaceable"><code>p</code></em> using maximum posterior decoding.</p>
	</td></tr><tr><td><p><span class="term">MAP-AU.html</span></p></td><td><p>An AU plot of the MAP alignment (AA/DNA color-cheme).
	</p></td></tr><tr><td><p><span class="term">P<em class="replaceable"><code>p</code></em>-max-AU.html</span></p></td><td>
	    <p>An AU plot of the maximum posterior decoding alignment for partition
	    <em class="replaceable"><code>p</code></em>  (AA/DNA color-cheme).</p>
	</td></tr></tbody></table></div>


      <p>The following files describe convergence and mixing:</p>


      <div class="variablelist"><table border="0" class="variablelist"><colgroup><col align="left" valign="top"><col></colgroup><tbody><tr><td><p><span class="term">partitions.bs</span></p></td><td>
	    <p>Confidence intervals on the support for partitions, generated
	      using a block bootstrap.</p>
	</td></tr><tr><td><p><span class="term">partitions.SRQ</span></p></td><td><p>A collection of
	      SRQ plots for the supported partitions.
	</p></td></tr><tr><td><p><span class="term">c50.SRQ</span></p></td><td><p>An
	      SRQ plot for the majority consensus tree.
	</p></td></tr></tbody></table></div>

      <p>The SRQ plots can be viewed by typing "<strong class="userinput"><code>plot
	  '<em class="replaceable"><code>file</code></em>' with lines</code></strong>" in
	<span class="application">gnuplot</span>.</p>

    </div>
      <div class="section"><div class="titlepage"><div><div><h4 class="title"><a name="idp58035200"></a>6.5.2. <code class="filename">Mixing/partitions.bs</code>: partition mixing</h4></div></div></div>
	
	<p>
	  This file reports the quality of estimates of support for each
	  partition in terms of the posterior probability (PP) and
	  log-10 odds (LOD).  It also reports the auto-correlation time (ACT),
	  the effective sample size (Ne), the number of samples
	  that support (1) or do not support (0) the partition, and
	  the number of regenerations. 

	  Only partitions with PP &gt; 0.1 are shown by default.
	</p>
      </div>



  </div>
  </div>
    

    <div class="section"><div class="titlepage"><div><div><h2 class="title" style="clear: both"><a name="subst_models"></a>7. Substitution models</h2></div></div></div>
      

       <p>Substitution models in
       <span class="application">BAli-Phy</span> are specified using a
       stack, as follows: <strong class="userinput"><code>Model[arg]+Model[arg]+...+Model[arg]</code></strong>
        where each model uses the previous models as input.  For
	example, <strong class="userinput"><code>LG+gamma_inv[4]</code></strong>. Arguments
	are optional.</p>

	<div class="note" style="margin-left: 0.5in; margin-right: 0.5in;"><h3 class="title">Note</h3>
	  <p>
	    If you are using the C-shell command line shell (<span class="command"><strong>csh</strong></span> or <span class="command"><strong>tcsh</strong></span>), then it will try to interpret each argument as an array reference, giving the error message "bali-phy: Not found."  To avoid this you may need to insert backslashes before the left square brackes, like this: <strong class="userinput"><code>Model\[arg]+Model\[arg]+...+Model\[arg]</code></strong>.
	  </p>
	</div>

      <div class="section"><div class="titlepage"><div><div><h3 class="title"><a name="idp58044000"></a>7.1. Default substitution models</h3></div></div></div>
	
	<p>If the substitution model is not specified, then the
	  default model for the alphabet is used.  For DNA or RNA, the
	  default model is TN.  For Triplets, the default is TNx3. For
          Codons, the default model is M0.  For Amino-Acids, the
	  default model is LG. 
	</p>
      </div>


      <div class="section"><div class="titlepage"><div><div><h3 class="title"><a name="subst_basic"></a>7.2. Basic CTMC models</h3></div></div></div>
	
	<p>The basic substitution models in
	  <span class="application">BAli-Phy</span> are continuous-time Markov
	  chains (CTMC). 
	  CTMC models can be characterized 
	  by transition rates <mml:math xmlns:mml="http://www.w3.org/1998/Math/MathML"><mml:msub><mml:mi>Q</mml:mi> <mml:mrow><mml:mi>ij</mml:mi></mml:mrow></mml:msub></mml:math> from letter <mml:math xmlns:mml="http://www.w3.org/1998/Math/MathML"><mml:mi>i</mml:mi></mml:math> to letter <mml:math xmlns:mml="http://www.w3.org/1998/Math/MathML"><mml:mi>j</mml:mi></mml:math>.
	  After a given time <mml:math xmlns:mml="http://www.w3.org/1998/Math/MathML"><mml:mi>t</mml:mi></mml:math> the probability for transition from state
	  <mml:math xmlns:mml="http://www.w3.org/1998/Math/MathML"><mml:mi>i</mml:mi></mml:math> to state <mml:math xmlns:mml="http://www.w3.org/1998/Math/MathML"><mml:mi>j</mml:mi></mml:math> is given by </p><div class="equation"><a name="idp58054352"></a><p class="title"><b>Equation 1. </b></p><div class="equation-contents"><mml:math xmlns:mml="http://www.w3.org/1998/Math/MathML"><mml:mi>P</mml:mi><mml:mo>(</mml:mo><mml:mi>t</mml:mi><mml:msub><mml:mo>)</mml:mo> <mml:mrow><mml:mi>ij</mml:mi></mml:mrow></mml:msub><mml:mo>=</mml:mo><mml:msup><mml:mi>e</mml:mi> <mml:mrow><mml:msub><mml:mi>Q</mml:mi> <mml:mrow><mml:mi>ij</mml:mi></mml:mrow></mml:msub><mml:mo>×</mml:mo><mml:mi>t</mml:mi></mml:mrow></mml:msup></mml:math></div></div><p><br class="equation-break">
	  using a matrix exponential.  Becase the CTMC models used in
	  <span class="application">BAli-Phy</span> are all reversible, the rate
	  matrix for these reversible models can be decomposed into a
	  symmetric matrix <mml:math xmlns:mml="http://www.w3.org/1998/Math/MathML"><mml:mi>S</mml:mi></mml:math> and equilibrium frequencies <mml:math xmlns:mml="http://www.w3.org/1998/Math/MathML"><mml:mi>&#960;</mml:mi></mml:math> as
	  follows:  </p><div class="equation"><a name="idp58061792"></a><p class="title"><b>Equation 2. </b></p><div class="equation-contents"><mml:math xmlns:mml="http://www.w3.org/1998/Math/MathML"><mml:msub><mml:mi>Q</mml:mi> <mml:mrow><mml:mi>ij</mml:mi></mml:mrow></mml:msub><mml:mo>=</mml:mo><mml:msub><mml:mi>S</mml:mi> <mml:mrow><mml:mi>ij</mml:mi></mml:mrow></mml:msub><mml:mo>×</mml:mo><mml:msub><mml:mi>&#960;</mml:mi> <mml:mi>j</mml:mi></mml:msub></mml:math></div></div><p><br class="equation-break">
	  The matrix <mml:math xmlns:mml="http://www.w3.org/1998/Math/MathML"><mml:mi>S</mml:mi></mml:math> is called the exchangability matrix, and
	  represents how exchangeable letters <mml:math xmlns:mml="http://www.w3.org/1998/Math/MathML"><mml:mi>i</mml:mi></mml:math> and <mml:math xmlns:mml="http://www.w3.org/1998/Math/MathML"><mml:mi>j</mml:mi></mml:math> are,
	  independent of their frequencies.
	</p>
	
	<p>The basic CTMC models are EQU, HKY, TN, GTR, HKYx3,
	TNx3, GTRx3, JTT, WAG, LG, and M0.  Each of these models is a
	way of specifying the exchangeability matrix <mml:math xmlns:mml="http://www.w3.org/1998/Math/MathML"><mml:msub><mml:mi>S</mml:mi> <mml:mrow><mml:mi>ij</mml:mi></mml:mrow></mml:msub></mml:math>.</p>

	<div class="table"><a name="idp58070384"></a><p class="title"><b>Table 1. Substitution Models</b></p><div class="table-contents">
	  
	  <table summary="Substitution Models" border="0"><colgroup><col class="col1"><col class="col2"><col class="col3"><col class="col4"></colgroup><thead><tr><th>Model</th><th>Alphabet</th><th>Parameters</th><th>Description</th></tr></thead><tbody><tr><td><strong class="userinput"><code>EQU</code></strong></td><td>any</td><td>none</td><td><mml:math xmlns:mml="http://www.w3.org/1998/Math/MathML"><mml:msub><mml:mi>S</mml:mi> <mml:mrow><mml:mi>ij</mml:mi></mml:mrow></mml:msub><mml:mo>=</mml:mo><mml:mn>1 </mml:mn></mml:math> for every <mml:math xmlns:mml="http://www.w3.org/1998/Math/MathML"><mml:mi>i</mml:mi></mml:math> and <mml:math xmlns:mml="http://www.w3.org/1998/Math/MathML"><mml:mi>j</mml:mi></mml:math>.</td></tr><tr><td>
		  <p><strong class="userinput"><code>HKY</code></strong></p>
		  <p>Hasegawa, Kishino, Yano (1985)</p>
		</td><td>DNA or RNA</td><td><mml:math xmlns:mml="http://www.w3.org/1998/Math/MathML"><mml:mi>&#954;</mml:mi></mml:math>: the ts/tv ratio.</td><td>
		  <p><mml:math xmlns:mml="http://www.w3.org/1998/Math/MathML"><mml:msub><mml:mi>S</mml:mi> <mml:mrow><mml:mi>ij</mml:mi></mml:mrow></mml:msub><mml:mo>=</mml:mo><mml:mn>1 </mml:mn></mml:math> for transversions.</p>
		  <p><mml:math xmlns:mml="http://www.w3.org/1998/Math/MathML"><mml:msub><mml:mi>S</mml:mi> <mml:mrow><mml:mi>ij</mml:mi></mml:mrow></mml:msub><mml:mo>=</mml:mo><mml:mi>&#954;</mml:mi></mml:math> for transitions.</p>
		</td></tr><tr><td><p><strong class="userinput"><code>TN</code></strong></p><p>Tamura, Nei (1993)</p></td><td>DNA or RNA</td><td>
		  <p><mml:math xmlns:mml="http://www.w3.org/1998/Math/MathML"><mml:msub><mml:mi>&#954;</mml:mi> <mml:mn>1 </mml:mn></mml:msub></mml:math>: the purine ts/tv ratio.</p>
		  <p><mml:math xmlns:mml="http://www.w3.org/1998/Math/MathML"><mml:msub><mml:mi>&#954;</mml:mi> <mml:mn>2 </mml:mn></mml:msub></mml:math>: the pyrimidine ts/tv ratio.</p>
		</td><td>
		  <p><mml:math xmlns:mml="http://www.w3.org/1998/Math/MathML"><mml:msub><mml:mi>S</mml:mi> <mml:mrow><mml:mi>ij</mml:mi></mml:mrow></mml:msub><mml:mo>=</mml:mo><mml:mn>1 </mml:mn></mml:math> for transversions.</p>
		  <p><mml:math xmlns:mml="http://www.w3.org/1998/Math/MathML"><mml:msub><mml:mi>S</mml:mi> <mml:mrow><mml:mi>ij</mml:mi></mml:mrow></mml:msub><mml:mo>=</mml:mo><mml:msub><mml:mi>&#954;</mml:mi> <mml:mn>1 </mml:mn></mml:msub></mml:math> for purine transitions.</p>
		  <p><mml:math xmlns:mml="http://www.w3.org/1998/Math/MathML"><mml:msub><mml:mi>S</mml:mi> <mml:mrow><mml:mi>ij</mml:mi></mml:mrow></mml:msub><mml:mo>=</mml:mo><mml:msub><mml:mi>&#954;</mml:mi> <mml:mn>2 </mml:mn></mml:msub></mml:math> for pyrimidine transitions.</p>
		</td></tr><tr><td><p><strong class="userinput"><code>GTR</code></strong></p><p>General Time-Reversible</p><p>Tavare (1986)</p></td><td>DNA or RNA</td><td>
		  <p><mml:math xmlns:mml="http://www.w3.org/1998/Math/MathML"><mml:msub><mml:mi>S</mml:mi> <mml:mrow><mml:mi>i</mml:mi><mml:mo>&#8800;</mml:mo><mml:mi>j</mml:mi></mml:mrow></mml:msub></mml:math></p>
		</td><td>
		  <p><mml:math xmlns:mml="http://www.w3.org/1998/Math/MathML"><mml:msub><mml:mo lspace="thinmathspace" rspace="thinmathspace">&#8721;</mml:mo> <mml:mrow><mml:mi>i</mml:mi><mml:mo>&#8800;</mml:mo><mml:mi>j</mml:mi></mml:mrow></mml:msub><mml:msub><mml:mi>S</mml:mi> <mml:mrow><mml:mi>ij</mml:mi></mml:mrow></mml:msub><mml:mo>=</mml:mo><mml:mn>1 </mml:mn></mml:math>.</p><p>(5 degrees of freedom).</p>

		</td></tr><tr><td>
		  <p><strong class="userinput"><code>JTT</code></strong></p>
		  <p>Jones, Taylor, Thornton (1992)</p>
		</td><td>Amino-Acids</td><td>
		  <p>none.</p>
		</td><td> </td></tr><tr><td>
		  <p><strong class="userinput"><code>WAG</code></strong></p>
		  <p>Whelan and Goldman (2001)</p>
		</td><td>Amino-Acids</td><td>
		  <p>none.</p>
		</td><td> </td></tr><tr><td>
		  <p><strong class="userinput"><code>LG</code></strong></p>
		  <p>Le and Gascuel (2008)</p>
		</td><td>Amino-Acids</td><td>
		  <p>none.</p>
		</td><td> </td></tr><tr><td>
		  <p><strong class="userinput"><code>Empirical[<em class="replaceable"><code>file</code></em>]</code></strong></p>
		</td><td>Amino-Acids</td><td>
		  <p>none.</p>
		</td><td>
		  <p>A user-specified empirical exchangeability matrix may be used.</p>
		  <p>The lower-triangular part of the symmetric
		  matrix is given, followed by the estimated
		  equilibrium frequencies.</p>
		</td></tr><tr><td>
		  <p><strong class="userinput"><code>HKYx3</code></strong></p>
		  <p><strong class="userinput"><code>TNx3</code></strong></p>
		  <p><strong class="userinput"><code>GTRx3</code></strong></p>
		</td><td>Triplets</td><td>
		  <p><em class="replaceable"><code>nuc-model</code></em> parameters.</p>
		</td><td>
		  <p>If the nuc-model has transition matrix
		    <mml:math xmlns:mml="http://www.w3.org/1998/Math/MathML"><mml:msubsup><mml:mi>S</mml:mi> <mml:mrow><mml:mi>ij</mml:mi></mml:mrow> <mml:mo>&#8242;</mml:mo></mml:msubsup></mml:math> on nucleotides, then:</p>

		  <p><mml:math xmlns:mml="http://www.w3.org/1998/Math/MathML"><mml:msub><mml:mi>S</mml:mi> <mml:mrow><mml:mi>&#945;</mml:mi><mml:mi>&#946;</mml:mi></mml:mrow></mml:msub><mml:mo>=</mml:mo><mml:mn>0 </mml:mn></mml:math> for changes of more than one nucleotide.</p>
		  <p><mml:math xmlns:mml="http://www.w3.org/1998/Math/MathML"><mml:msub><mml:mi>S</mml:mi> <mml:mrow><mml:mi>&#945;</mml:mi><mml:mi>&#946;</mml:mi></mml:mrow></mml:msub><mml:mo>=</mml:mo><mml:msubsup><mml:mi>S</mml:mi> <mml:mrow><mml:mi>ij</mml:mi></mml:mrow> <mml:mo>&#8242;</mml:mo></mml:msubsup></mml:math> for single nucleotide  changes <mml:math xmlns:mml="http://www.w3.org/1998/Math/MathML"><mml:mi>i</mml:mi><mml:mo>&#8594;</mml:mo><mml:mi>j</mml:mi></mml:math>.</p>
		</td></tr><tr><td>
		  <p><strong class="userinput"><code>M0</code></strong></p>
		  <p>Nielsen and Yang (1998)</p>
		</td><td>Codons</td><td>
		  <p><mml:math xmlns:mml="http://www.w3.org/1998/Math/MathML"><mml:mi>&#954;</mml:mi></mml:math>: the ts/tv ratio.</p>
		  <p><mml:math xmlns:mml="http://www.w3.org/1998/Math/MathML"><mml:mi>&#969;</mml:mi></mml:math>: the dN/dS ratio.</p>
		</td><td>
		  <p><mml:math xmlns:mml="http://www.w3.org/1998/Math/MathML"><mml:msub><mml:mi>S</mml:mi> <mml:mrow><mml:mi>&#945;</mml:mi><mml:mi>&#946;</mml:mi></mml:mrow></mml:msub><mml:mo>=</mml:mo><mml:mn>0 </mml:mn></mml:math> for changes of more than one nucleotide.</p>
		  <p><mml:math xmlns:mml="http://www.w3.org/1998/Math/MathML"><mml:msub><mml:mi>S</mml:mi> <mml:mrow><mml:mi>&#945;</mml:mi><mml:mi>&#946;</mml:mi></mml:mrow></mml:msub><mml:mo>=</mml:mo><mml:mn>1 </mml:mn></mml:math> for synonymous transversions.</p>
		  <p><mml:math xmlns:mml="http://www.w3.org/1998/Math/MathML"><mml:msub><mml:mi>S</mml:mi> <mml:mrow><mml:mi>&#945;</mml:mi><mml:mi>&#946;</mml:mi></mml:mrow></mml:msub><mml:mo>=</mml:mo><mml:mi>&#969;</mml:mi></mml:math> for non-synonymous transversions.</p>

		  <p><mml:math xmlns:mml="http://www.w3.org/1998/Math/MathML"><mml:msub><mml:mi>S</mml:mi> <mml:mrow><mml:mi>&#945;</mml:mi><mml:mi>&#946;</mml:mi></mml:mrow></mml:msub><mml:mo>=</mml:mo><mml:mi>&#954;</mml:mi></mml:math> for synonymous transitions.</p>
		  <p><mml:math xmlns:mml="http://www.w3.org/1998/Math/MathML"><mml:msub><mml:mi>S</mml:mi> <mml:mrow><mml:mi>&#945;</mml:mi><mml:mi>&#946;</mml:mi></mml:mrow></mml:msub><mml:mo>=</mml:mo><mml:mi>&#969;</mml:mi><mml:mi>&#954;</mml:mi></mml:math> for non-synonymous transitions.</p>
		</td></tr><tr><td>
		  <p><strong class="userinput"><code>M0[<em class="replaceable"><code>nuc-model=HKY</code></em>]</code></strong></p>
		  <p>Nielsen and Yang (1998)</p>
		</td><td>Codons</td><td>
		  <p><em class="replaceable"><code>nuc-model</code></em> parameters.</p>
		  <p><mml:math xmlns:mml="http://www.w3.org/1998/Math/MathML"><mml:mi>&#969;</mml:mi></mml:math>: the dn/ds ratio</p>
		</td><td>
		  <p>If the nuc-model has transition matrix
		    <mml:math xmlns:mml="http://www.w3.org/1998/Math/MathML"><mml:msubsup><mml:mi>S</mml:mi> <mml:mrow><mml:mi>ij</mml:mi></mml:mrow> <mml:mo>&#8242;</mml:mo></mml:msubsup></mml:math> on nucleotides, then:</p>

		  <p><mml:math xmlns:mml="http://www.w3.org/1998/Math/MathML"><mml:msub><mml:mi>S</mml:mi> <mml:mrow><mml:mi>&#945;</mml:mi><mml:mi>&#946;</mml:mi></mml:mrow></mml:msub><mml:mo>=</mml:mo><mml:mn>0 </mml:mn></mml:math> for changes of more than one nucleotide.</p>
		  <p><mml:math xmlns:mml="http://www.w3.org/1998/Math/MathML"><mml:msub><mml:mi>S</mml:mi> <mml:mrow><mml:mi>&#945;</mml:mi><mml:mi>&#946;</mml:mi></mml:mrow></mml:msub><mml:mo>=</mml:mo><mml:msubsup><mml:mi>S</mml:mi> <mml:mrow><mml:mi>ij</mml:mi></mml:mrow> <mml:mo>&#8242;</mml:mo></mml:msubsup></mml:math> for synonymous changes.</p>
		  <p><mml:math xmlns:mml="http://www.w3.org/1998/Math/MathML"><mml:msub><mml:mi>S</mml:mi> <mml:mrow><mml:mi>&#945;</mml:mi><mml:mi>&#946;</mml:mi></mml:mrow></mml:msub><mml:mo>=</mml:mo><mml:mi>&#969;</mml:mi><mml:msubsup><mml:mi>S</mml:mi> <mml:mrow><mml:mi>ij</mml:mi></mml:mrow> <mml:mo>&#8242;</mml:mo></mml:msubsup></mml:math> for non-synonymous changes.</p>
		</td></tr></tbody></table>
	</div></div><br class="table-break">

      </div>

    <div class="section"><div class="titlepage"><div><div><h3 class="title"><a name="subst_freq"></a>7.3. Substitution Frequency models</h3></div></div></div>
      
      <p>The rate matrix <mml:math xmlns:mml="http://www.w3.org/1998/Math/MathML"><mml:msub><mml:mi>Q</mml:mi> <mml:mrow><mml:mi>ij</mml:mi></mml:mrow></mml:msub></mml:math> can be more generally expressed as
	</p><div class="equation"><a name="idp58192416"></a><p class="title"><b>Equation 3. </b></p><div class="equation-contents"><mml:math xmlns:mml="http://www.w3.org/1998/Math/MathML"><mml:msub><mml:mi>Q</mml:mi> <mml:mrow><mml:mi>ij</mml:mi></mml:mrow></mml:msub><mml:mo>=</mml:mo><mml:msub><mml:mi>S</mml:mi> <mml:mrow><mml:mi>ij</mml:mi></mml:mrow></mml:msub><mml:mo>×</mml:mo><mml:mfrac><mml:mrow><mml:msubsup><mml:mi>&#960;</mml:mi> <mml:mi>j</mml:mi> <mml:mi>f</mml:mi></mml:msubsup></mml:mrow><mml:mrow><mml:msubsup><mml:mi>&#960;</mml:mi> <mml:mi>i</mml:mi> <mml:mrow><mml:mn>1 </mml:mn><mml:mo>-</mml:mo><mml:mi>f</mml:mi></mml:mrow></mml:msubsup></mml:mrow></mml:mfrac><mml:mo>,</mml:mo></mml:math></div></div><p><br class="equation-break">
        where <mml:math xmlns:mml="http://www.w3.org/1998/Math/MathML"><mml:mi>f</mml:mi></mml:math> ranges from <mml:math xmlns:mml="http://www.w3.org/1998/Math/MathML"><mml:mn>0 </mml:mn></mml:math> to <mml:math xmlns:mml="http://www.w3.org/1998/Math/MathML"><mml:mn>1 </mml:mn></mml:math>.
	Here the parameter <mml:math xmlns:mml="http://www.w3.org/1998/Math/MathML"><mml:mi>f</mml:mi></mml:math> specifies the relative importance of
	unequal conservation (<mml:math xmlns:mml="http://www.w3.org/1998/Math/MathML"><mml:mi>f</mml:mi><mml:mo>=</mml:mo><mml:mn>0 </mml:mn></mml:math>) and unequal replacement (<mml:math xmlns:mml="http://www.w3.org/1998/Math/MathML"><mml:mi>f</mml:mi><mml:mo>=</mml:mo><mml:mn>1 </mml:mn></mml:math>) in
	maintaining the equilibrium frequencies <mml:math xmlns:mml="http://www.w3.org/1998/Math/MathML"><mml:mi>&#960;</mml:mi></mml:math>.  
      </p>
      <p>
	In fact, this can be generalized even further to </p><div class="equation"><a name="idp58207056"></a><p class="title"><b>Equation 4. </b></p><div class="equation-contents"><mml:math xmlns:mml="http://www.w3.org/1998/Math/MathML"><mml:msub><mml:mi>Q</mml:mi> <mml:mrow><mml:mi>ij</mml:mi></mml:mrow></mml:msub><mml:mo>=</mml:mo><mml:msub><mml:mi>S</mml:mi> <mml:mrow><mml:mi>ij</mml:mi></mml:mrow></mml:msub><mml:mo>×</mml:mo><mml:mi>R</mml:mi><mml:mo>(</mml:mo><mml:mi>&#960;</mml:mi><mml:msub><mml:mo>)</mml:mo> <mml:mrow><mml:mi>ij</mml:mi></mml:mrow></mml:msub></mml:math></div></div><p><br class="equation-break"> where </p><div class="equation"><a name="idp58212048"></a><p class="title"><b>Equation 5. </b></p><div class="equation-contents"><mml:math xmlns:mml="http://www.w3.org/1998/Math/MathML"><mml:msub><mml:mi>&#960;</mml:mi> <mml:mi>i</mml:mi></mml:msub><mml:mo>×</mml:mo><mml:msub><mml:mi>R</mml:mi> <mml:mrow><mml:mi>ij</mml:mi></mml:mrow></mml:msub><mml:mo>=</mml:mo><mml:msub><mml:mi>&#960;</mml:mi> <mml:mi>j</mml:mi></mml:msub><mml:mo>×</mml:mo><mml:msub><mml:mi>R</mml:mi> <mml:mrow><mml:mi>ji</mml:mi></mml:mrow></mml:msub><mml:mo>.</mml:mo></mml:math></div></div><p><br class="equation-break">
      </p>
      <p>These models can therefore be expressed as a combination
	of an "exchange model" (for <mml:math xmlns:mml="http://www.w3.org/1998/Math/MathML"><mml:mi>S</mml:mi></mml:math>) and a "frequency model" (for
	<mml:math xmlns:mml="http://www.w3.org/1998/Math/MathML"><mml:mi>R</mml:mi></mml:math>).
      </p>

      <div class="table"><a name="idp58219408"></a><p class="title"><b>Table 2. Frequency Models</b></p><div class="table-contents">
	
	<table summary="Frequency Models" border="0"><colgroup><col class="col1"><col class="col2"><col class="col3"></colgroup><thead><tr><th>Model</th><th>Alphabet</th><th>Parameters</th></tr></thead><tbody><tr><td>
		<p><strong class="userinput"><code>F</code></strong></p>
		<p>Whelan and Goldman (2001)</p>
	      </td><td>any</td><td>
		<p>pi<em class="replaceable"><code>l</code></em>: frequency of letter <em class="replaceable"><code>l</code></em>.</p>
	      </td></tr><tr><td>
		<p><strong class="userinput"><code>F=constant</code></strong></p>
		<p>Whelan and Goldman (2001)</p>
		<p>Frequencies fixed to simple estimates.</p>
	      </td><td>any</td><td>
	      </td></tr><tr><td>
		<p><strong class="userinput"><code>F=uniform</code></strong></p>
		<p>Whelan and Goldman (2001)</p>
		<p>All letters have equal frequencies.</p>
	      </td><td>any</td><td>
	      </td></tr><tr><td>
		<p><strong class="userinput"><code>F61</code></strong></p>
		<p>Synonym of F for codon models</p>
	      </td><td>any</td><td>
		<p>pi<em class="replaceable"><code>l</code></em>: frequency of letter <em class="replaceable"><code>l</code></em>.</p>
	      </td></tr><tr><td>
		<p><strong class="userinput"><code>gwF</code></strong></p>
		<p>Goldman and Whelan (2002)</p>
	      </td><td>any</td><td>
		<p>f: determines cause of high-frequency letters.</p>
		<p>pi<em class="replaceable"><code>l</code></em>: frequency of letter <em class="replaceable"><code>l</code></em>.</p>
	      </td></tr><tr><td>
		<p><strong class="userinput"><code>F1x4</code></strong></p>
		<p>Single nucleotide frequency model</p>
	      </td><td>Triplets</td><td>
		<p>pi<em class="replaceable"><code>l</code></em>: frequency
		of nucleotide <em class="replaceable"><code>l</code></em> in first codon position.</p>
	      </td></tr><tr><td>
		<p><strong class="userinput"><code>F3x4</code></strong></p>
		<p>Independent nucleotide frequency model</p>
	      </td><td>Triplets</td><td>
		<p>Site1.pi<em class="replaceable"><code>l</code></em>: frequency
		of letter <em class="replaceable"><code>l</code></em> in first codon position.</p>
		<p>Site2.pi<em class="replaceable"><code>l</code></em>: frequency
		of letter <em class="replaceable"><code>l</code></em> in first codon position.</p>
		<p>Site3.pi<em class="replaceable"><code>l</code></em>: frequency
		of letter <em class="replaceable"><code>l</code></em> in first codon position.</p>
	      </td></tr><tr><td>
		<p><strong class="userinput"><code>MG94</code></strong></p>
		<p>Muse and Gaut (1994)</p>
		<p>Single nucleotide frequency model</p>
	      </td><td>Triplets</td><td>
		<p>pi<em class="replaceable"><code>l</code></em>: frequency
		of nucleotide <em class="replaceable"><code>l</code></em> in first codon position.</p>
	      </td></tr><tr><td>
		<p><strong class="userinput"><code>MG94w9</code></strong></p>
		<p>Muse and Gaut (1994)</p>
		<p>Independent nucleotide frequency model</p>
	      </td><td>Triplets</td><td>
		<p>Site1.pi<em class="replaceable"><code>l</code></em>: frequency
		of letter <em class="replaceable"><code>l</code></em> in first codon position.</p>
		<p>Site2.pi<em class="replaceable"><code>l</code></em>: frequency
		of letter <em class="replaceable"><code>l</code></em> in first codon position.</p>
		<p>Site3.pi<em class="replaceable"><code>l</code></em>: frequency
		of letter <em class="replaceable"><code>l</code></em> in first codon position.</p>
	      </td></tr></tbody></table>
      </div></div><br class="table-break">
    </div>


    <div class="section"><div class="titlepage"><div><div><h3 class="title"><a name="idp58274288"></a>7.4. Substitution Mixture Models</h3></div></div></div>
    
    <p>
      Complex substitution models in <span class="application">BAli-Phy</span>
      are constructed as mixtures of reversible CTMC models (see <a class="xref" href="#subst_basic" title="7.2. Basic CTMC models">Section 7.2, &#8220;Basic CTMC models&#8221;</a>) that run at different rates (e.g. <mml:math xmlns:mml="http://www.w3.org/1998/Math/MathML"><mml:msub><mml:mi>&#915;</mml:mi> <mml:mn>4 </mml:mn></mml:msub><mml:mo>+</mml:mo><mml:mi>INV</mml:mi></mml:math>)
      or have different parameters (e.g. an M2 codon model).
    </p>

      <p>Model modifiers are gamma, log-normal, INV, M2, M3, and M7.</p>


      <div class="table"><a name="idp58279312"></a><p class="title"><b>Table 3. CTMC Mixture Models</b></p><div class="table-contents">
	
	<table summary="CTMC Mixture Models" border="0"><colgroup><col><col><col></colgroup><thead><tr><th>Model</th><th>Alphabet</th><th>Parameters</th></tr></thead><tbody><tr><td>
		<p><strong class="userinput"><code><em class="replaceable"><code>sub-model</code></em> + gamma</code></strong></p>
		<p><strong class="userinput"><code><em class="replaceable"><code>sub-model</code></em> + gamma[4]</code></strong> (default options)</p>
		<p><strong class="userinput"><code><em class="replaceable"><code>sub-model</code></em> + gamma[<em class="replaceable"><code>n</code></em>]</code></strong></p>
		<p>Yang (1994)</p>
	      </td><td><em class="replaceable"><code>sub-model</code></em> alphabet</td><td>
		<p>sigmaOverMu: noise to signal ratio for <mml:math xmlns:mml="http://www.w3.org/1998/Math/MathML"><mml:mi>&#915;</mml:mi></mml:math>.  Also equal to 1/sqrt(<mml:math xmlns:mml="http://www.w3.org/1998/Math/MathML"><mml:mi>&#945;</mml:mi></mml:math>).</p>
		<p><em class="replaceable"><code>n</code></em>: Number of bins for discrete <mml:math xmlns:mml="http://www.w3.org/1998/Math/MathML"><mml:mi>&#915;</mml:mi></mml:math>.</p>
	      </td></tr><tr><td>
		<p><strong class="userinput"><code><em class="replaceable"><code>sub-model</code></em> + gamma_inv</code></strong></p>
		<p><strong class="userinput"><code><em class="replaceable"><code>sub-model</code></em> + gamma_inv[4]</code></strong> (default options)</p>
		<p><strong class="userinput"><code><em class="replaceable"><code>sub-model</code></em> + gamma_inv[<em class="replaceable"><code>n</code></em>]</code></strong></p>
		<p>Yang (1994)</p>
	      </td><td><em class="replaceable"><code>sub-model</code></em> alphabet</td><td>
		<p>sigmaOverMu: noise to signal ratio for <mml:math xmlns:mml="http://www.w3.org/1998/Math/MathML"><mml:mi>&#915;</mml:mi></mml:math>.  Also equal to 1/sqrt(<mml:math xmlns:mml="http://www.w3.org/1998/Math/MathML"><mml:mi>&#945;</mml:mi></mml:math>).</p>
		<p><em class="replaceable"><code>n</code></em>: Number of bins for discrete <mml:math xmlns:mml="http://www.w3.org/1998/Math/MathML"><mml:mi>&#915;</mml:mi></mml:math>.</p>
		<p><mml:math xmlns:mml="http://www.w3.org/1998/Math/MathML"><mml:mi>p</mml:mi></mml:math>: Fraction of invariant sites.</p>
	      </td></tr><tr><td>
		<p><strong class="userinput"><code><em class="replaceable"><code>sub-model</code></em> + log-normal</code></strong></p>
		<p><strong class="userinput"><code><em class="replaceable"><code>sub-model</code></em> + log-normal[4]</code></strong> (default options)</p>
		<p><strong class="userinput"><code><em class="replaceable"><code>sub-model</code></em> + log-normal[<em class="replaceable"><code>n</code></em>]</code></strong></p>
	      </td><td><em class="replaceable"><code>sub-model</code></em> alphabet</td><td>
		<p>sigmaOverMu: <mml:math xmlns:mml="http://www.w3.org/1998/Math/MathML"><mml:mi>&#963;</mml:mi><mml:mo>/</mml:mo><mml:mi>&#956;</mml:mi></mml:math>: noise to signal ratio for <mml:math xmlns:mml="http://www.w3.org/1998/Math/MathML"><mml:mi>log</mml:mi><mml:mi>Normal</mml:mi></mml:math>.</p>
		<p><em class="replaceable"><code>n</code></em>: Number of bins for discrete <mml:math xmlns:mml="http://www.w3.org/1998/Math/MathML"><mml:mi>log</mml:mi><mml:mi>Normal</mml:mi></mml:math>.</p>
		<p><mml:math xmlns:mml="http://www.w3.org/1998/Math/MathML"><mml:mi>p</mml:mi></mml:math>: Fraction of invariant sites.</p>
	      </td></tr><tr><td>
		<p><strong class="userinput"><code><em class="replaceable"><code>sub-model</code></em> + log-normal_inv</code></strong></p>
		<p><strong class="userinput"><code><em class="replaceable"><code>sub-model</code></em> + log-normal_inv[4]</code></strong> (default options)</p>
		<p><strong class="userinput"><code><em class="replaceable"><code>sub-model</code></em> + log-normal_inv[<em class="replaceable"><code>n</code></em>]</code></strong></p>
	      </td><td><em class="replaceable"><code>sub-model</code></em> alphabet</td><td>
		<p><mml:math xmlns:mml="http://www.w3.org/1998/Math/MathML"><mml:mi>&#963;</mml:mi><mml:mo>/</mml:mo><mml:mi>&#956;</mml:mi></mml:math>: noise to signal ratio for <mml:math xmlns:mml="http://www.w3.org/1998/Math/MathML"><mml:mi>log</mml:mi><mml:mi>Normal</mml:mi></mml:math>.</p>
		<p><em class="replaceable"><code>n</code></em>: Number of bins for discrete <mml:math xmlns:mml="http://www.w3.org/1998/Math/MathML"><mml:mi>log</mml:mi><mml:mi>Normal</mml:mi></mml:math>.</p>
		<p><mml:math xmlns:mml="http://www.w3.org/1998/Math/MathML"><mml:mi>p</mml:mi></mml:math>: Fraction of invariant sites.</p>
	      </td></tr><tr><td>
		<p><strong class="userinput"><code>M1a</code></strong></p>
		<p><strong class="userinput"><code>M1a[HKY,gwF]</code></strong> (default options)</p>
		<p><strong class="userinput"><code>M1a[GTR,F3x4]</code></strong></p>
		<p>Wong, et. al. (2004)</p>
	      </td><td>Codons</td><td>
		<p>kappa: the transition/transversion ratio</p>
		<p>omega1: the conserved dN/dS ratio</p>
		<p>fConserved: the fraction of conserved sites</p>
		<p>fNeutral: the fraction of neutral sites</p>
	      </td></tr><tr><td>
		<p><strong class="userinput"><code>M2a</code></strong></p>
		<p><strong class="userinput"><code>M2a[HKY,gwF]</code></strong> (default options)</p>
		<p><strong class="userinput"><code>M2a[GTR,F3x4]</code></strong></p>
		<p>Wong, et. al. (2004)</p>
	      </td><td>Codons</td><td>
		<p>kappa: the transition/transversion ratio</p>
		<p>omega1: the conserved dN/dS ratio</p>
		<p>omega3: the positively selected dN/dS ratio</p>
		<p>fConserved: the fraction of conserved sites</p>
		<p>fNeutral: the fraction of neutral sites</p>
		<p>fSelection: the fraction of neutral sites</p>
	      </td></tr><tr><td>
		<p><strong class="userinput"><code>M2a_Test</code></strong></p>
		<p><strong class="userinput"><code>M2a_Test[HKY,gwF]</code></strong> (default options)</p>
		<p><strong class="userinput"><code>M2a_Test[GTR,F3x4]</code></strong></p>
		<p>Wong, et. al. (2004)</p>
	      </td><td>Codons</td><td>
		<p>kappa: the transition/transversion ratio</p>
		<p>omega1: the conserved dN/dS ratio</p>
		<p>omega3: the positively selected dN/dS ratio</p>
		<p>fConserved: the fraction of conserved sites</p>
		<p>fNeutral: the fraction of neutral sites</p>
		<p>fSelection: the fraction of neutral sites</p>
		<p>pos_selection: model selector (1 if positive selection, 0 if not).</p>
	      </td></tr><tr><td>
		<p><strong class="userinput"><code>branch-site</code></strong></p>
		<p><strong class="userinput"><code>branch-site[2,HKY,gwF]</code></strong>(default options)</p>
		<p><strong class="userinput"><code>branch-site[,HKY,F3x4]</code></strong></p>
		<p><strong class="userinput"><code>branch-site[3,GTR,F61]</code></strong></p>
		<p>Zhang, et. al. (2005)</p>
	      </td><td>Codons</td><td>
		<p>f0: the fraction of sites conserved on background branches.</p>
		<p>f1: the fraction of sites neutral on background branches.</p>
		<p>posP: the fraction of sites (neutral or
		conserved) that switch to positive selection on
		foreground branches.</p>
		<p>w0: the conserved dN/dS ratio.</p>
		<p>posW: the positively selected dN/dS ratio.</p>
		<p>posSelection: model selector (1 if positive selection, 0 if not).</p>
	      </td></tr></tbody></table>
      </div></div><br class="table-break">
    </div>

    <div class="section"><div class="titlepage"><div><div><h3 class="title"><a name="idp58362320"></a>7.5. The branch-site substitution model</h3></div></div></div>
    <p>In order to use the branch-site substitution model, the user needs to
    </p><div class="itemizedlist"><ul class="itemizedlist" style="list-style-type: disc; "><li class="listitem">Specify an unrooted tree topology in Newick format.</li><li class="listitem">Label foreground branches on the tree using NHX attributes.</li><li class="listitem">Disable topology changes in order to fix the tree topology (branch lengths are estimated).</li></ul></div><p>
    Here is an example tree file:
</p><div class="example"><a name="idp58365568"></a><p class="title"><b>Example 4. An initial tree file with branch lengths</b></p><div class="example-contents">
(((A1:0.1, B1:0.1):0.1,(C1:0.1, D1:0.1):0.1):0.1,((E1:0.1, F1:0.1):0.1,(G1:0.1, H1:0.1):0.1):0.1,(((A2:0.1, B2:0.1):0.1,(C2:0.1, D2:0.1):0.1):0.1,((E2:0.1, F2:0.1):0.1,(G2:0.1, H2:0.1):0.1):0.1):[&amp;&amp;NHX:foreground=1]0.2);
</div></div><p><br class="example-break">
Any branch lengths provided will be used as initial values in the MCMC analysis.  However, it is not necessary to provide them:
</p><div class="example"><a name="idp58366656"></a><p class="title"><b>Example 5. An initial tree file without branch lengths</b></p><div class="example-contents">
(((A1, B1),(C1, D1)),((E1, F1),(G1, H1)),(((A2, B2),(C2, D2)),((E2, F2),(G2, H2))):[&amp;&amp;NHX:foreground=1]);
</div></div><p><br class="example-break">
The NHX attribute must be applied to the branch, not the node.  Therefore it must occur after a colon.  Multiple branches may be marked as foreground branches.
    </p>
    <p>An example command line is as follows:
</p><pre class="screen"><code class="prompt">%</code> bali-phy <em class="replaceable"><code>alignment</code></em>.fasta --smodel branch-site[,HKY,F3x4] --disable=topology --tree=<em class="replaceable"><code>tree</code></em>.tree --alphabet=Codons</pre><p>
</p>
<p>The posterior probability of positive selection is the posterior mean of the posSelection parameter.  This may be computed using the statreport program with the <strong class="userinput"><code>--mean</code></strong> option.
</p>
<p>In case this probability is extremely close to 1 or 0, you may wish to add the option <strong class="userinput"><code>--Rao-Blackwellize S1.BranchSite.posSelection</code></strong>.  This will report the log-probability of positive selection each iteration.  The user may exponentiate the reported values and then average them (using R, for example) in order to compute a more accurate estimate of the posterior probability of positive selection.
    </p>
    </div>
    <div class="section"><div class="titlepage"><div><div><h3 class="title"><a name="idp58372192"></a>7.6. Substitution model examples</h3></div></div></div>
      
      <p>Example: --smodel WAG+F+log-normal_inv</p>
      <p>Example: --smodel WAG+gwF+log-normal_inv</p>
      <p>Example: --smodel WAG+log-normal_inv (same as above)</p>
      <p>Example: --smodel EQU --alphabet Triplets</p>
      <p>Example: --smodel HKY</p>
      <p>Example: --smodel TN+F=constant</p>
      <p>Example: --smodel M0 --alphabet Codons</p>
      <p>Example: --smodel M0+F1x4 --alphabet Codons</p>
      <p>Example: --smodel M2a --alphabet Codons</p>
      <p>Example: --smodel M2a[HKY] --alphabet Codons (same as above)</p>
      <p>Example: --smodel M2[TN,F3x4] --alphabet Codons</p>
    </div>
    <div class="section"><div class="titlepage"><div><div><h3 class="title"><a name="genetic-codes"></a>7.7. Genetic Codes</h3></div></div></div>
      
      <p>When using a codon-based substitution model like <strong class="userinput"><code>M0</code></strong>, you may select the genetic code by specifying <strong class="userinput"><code>--alphabet Codons[<em class="replaceable"><code>genetic-code</code></em>]</code></strong>.  Available genetic codes are <strong class="userinput"><code>standard</code></strong>, <strong class="userinput"><code>mt-vert</code></strong>, <strong class="userinput"><code>mt-invert</code></strong>, <strong class="userinput"><code>mt-yeast</code></strong>, <strong class="userinput"><code>mt-protozoan</code></strong>.</p>
<p>If the genetic code is not specified, then the standard code is used:
	</p><pre class="screen"><code class="prompt">%</code> bali-phy <em class="replaceable"><code>sequence-file</code></em> --smodel M0 --alphabet Codons</pre><p>
	This example specifies the vertebrate mitochondrial code:
	</p><pre class="screen"><code class="prompt">%</code> bali-phy <em class="replaceable"><code>sequence-file</code></em> --smodel M0 --alphabet Codons[mt-vert]</pre><p>
</p>
    </div>

  </div>

    <div class="section"><div class="titlepage"><div><div><h2 class="title" style="clear: both"><a name="indel_models"></a>8. Insertion/deletion models</h2></div></div></div>
      
      <p>The current models are RS05, RS07, and <strong class="userinput"><code>none</code></strong>.  The default is RS07.  Each of these models is a probability distribution on pairwise alignments.  The probability distribution on multiple sequence alignments <mml:math xmlns:mml="http://www.w3.org/1998/Math/MathML"><mml:mi>Pr</mml:mi><mml:mo>(</mml:mo><mml:mi>A</mml:mi><mml:mo lspace="0em" rspace="0em">&#8739;</mml:mo><mml:mi>T</mml:mi><mml:mo>,</mml:mo><mml:mi>&#964;</mml:mi><mml:mo>,</mml:mo><mml:mi>&#923;</mml:mi><mml:mo>)</mml:mo></mml:math> is constructed by factoring the multiple sequence alignment into pairwise alignments along each branch of the tree, as described in Redelings and Suchard (2005).</p>

<div class="table"><a name="idp58393904"></a><p class="title"><b>Table 4. Substitution Models</b></p><div class="table-contents">
	  
	  <table summary="Substitution Models" border="0"><colgroup><col class="col1"><col class="col2"><col class="col3"></colgroup><thead><tr><th>Model</th><th>Parameters</th><th>Description</th></tr></thead><tbody><tr><td>
		  <p><strong class="userinput"><code>RS05</code></strong></p>
		  <p>Redelings and Suchard (2005)</p>
		</td><td>
		  <p><mml:math xmlns:mml="http://www.w3.org/1998/Math/MathML"><mml:mi>&#948;</mml:mi></mml:math>: the gap-opening probability</p>
		  <p><mml:math xmlns:mml="http://www.w3.org/1998/Math/MathML"><mml:mi>&#8714;</mml:mi></mml:math>: the gap-extension probability</p>
		</td><td>
		  <p>Gap lengths are geometrically distributed with extension probability <mml:math xmlns:mml="http://www.w3.org/1998/Math/MathML"><mml:mi>&#8714;</mml:mi></mml:math>.</p>
		  <p>This indel model is independent of the branch length connecting the ancestor and descent sequences.</p>
		</td></tr><tr><td>
		  <p><strong class="userinput"><code>RS07</code></strong></p>
		  <p>Redelings and Suchard (2007)</p>
		</td><td>
		  <p><mml:math xmlns:mml="http://www.w3.org/1998/Math/MathML"><mml:mi>&#955;</mml:mi></mml:math>: the insertion and deletion rate</p>
		  <p><mml:math xmlns:mml="http://www.w3.org/1998/Math/MathML"><mml:mi>&#8714;</mml:mi></mml:math>: the gap-extension probability</p>
		</td><td>
		  <p>Gap lengths are geometrically distributed with extension probability <mml:math xmlns:mml="http://www.w3.org/1998/Math/MathML"><mml:mi>&#8714;</mml:mi></mml:math>.</p>
		  <p>This probability of an indel event depends on the branch length in this model.</p>
		</td></tr><tr><td>
		  <p><strong class="userinput"><code>none</code></strong></p>
		</td><td> </td><td>	Indicates the lack of a model.</td></tr></tbody></table>
	</div></div><br class="table-break">
	    

      <p>Specifying an indel model of <strong class="userinput"><code>none</code></strong> for a given partition results in fixing the alignment for that partition to its initial value, and ignoring information in shared insertions or deletions.</p>
    </div>

    <div class="section"><div class="titlepage"><div><div><h2 class="title" style="clear: both"><a name="alignment_constraints"></a>9. Alignment constraints</h2></div></div></div>
      
      <p>To fix specific columns of the alignment, you may specify alignment
	constraints in a file as follows: </p>

      <div class="orderedlist"><ol class="orderedlist" type="1"><li class="listitem"><p>Use the argument
	    <strong class="userinput"><code>--align-constraint <em class="replaceable"><code>filename</code></em></code></strong></p></li><li class="listitem"><p>The filename refers to a file in which each
	    line represents a constraint.</p></li></ol></div>

      <div class="section"><div class="titlepage"><div><div><h3 class="title"><a name="idp58424528"></a>9.1. Syntax</h3></div></div></div>
	
	<p>The first line of the file is a header consisting of an
	  ordered list of sequence names separated by spaces. Each subsequent line
	  consists of a space-separated list of sequence positions, with the first position
	  corresponding to the first leaf sequence, the second position
	  corresponding to the second leaf sequence, etc. Thus, if there are
	  <em class="replaceable"><code>n</code></em> leaf taxa, then each line corresponds to a
	  space-separated list of <em class="replaceable"><code>n</code></em> integers.</p>
      </div>
      <div class="section"><div class="titlepage"><div><div><h3 class="title"><a name="idp58427312"></a>9.2. Examples</h3></div></div></div>
	
	<p>For example, the file</p>

	<pre class="programlisting">A B C
1 2 2</pre>

	<p>implies that position 1 of leaf sequence A is aligned to
	  position 2 of leaf sequences B and C. Note that the first
	  position in a sequence is position 0.</p> 

	<p>Optionally, one may use a '-' instead of an integer, which
	  denotes a lack of constraint for that sequence. This can be
	  useful as follows:</p> 

	<pre class="programlisting">A B C D
2 2 - - 
- - 2 2
	</pre>

	<p>The above constraints force alignment between position 2
	  of sequences A and B, and between position 2 of sequence C and
	  D.</p>
      </div>

      <div class="section"><div class="titlepage"><div><div><h3 class="title"><a name="idp58431584"></a>9.3. Computing the constraints</h3></div></div></div>
	
	<p>
	  The program <span class="command"><strong>alignment-indices</strong></span> may be used
	  to aid in computing a constraint file from an input
	  alignment.  See <a class="xref" href="#generating_constraint_files" title="12.6.3.">Q: 12.6.3</a>.
	</p>
      </div>
    </div>

  <div class="section"><div class="titlepage"><div><div><h2 class="title" style="clear: both"><a name="mixing_and_convergence"></a>10. Convergence and Mixing: Is it done yet?</h2></div></div></div>
    

    <p>
      When using Markov chain Monte Carlo (MCMC) programs like
      <span class="application">MrBayes</span>, <span class="application">BEAST</span> or
      <span class="application">BAli-Phy</span>, it is hard to determine in
      advance how many iterations are required to give a good
      estimate. The number depends on the specific data set that is
      being examined. As a result, <span class="application">BAli-Phy</span>
      relies on the user to analyze the output of a running chain
      periodically in order to determine when enough samples have been
      obtained.  This section describes a number of techniques to
      diagnose when more samples must be taken.
    </p>

    <p>Some of the better diagnostics for lack of convergence rely on running at least 4 independent copies of the Markov chain (preferably 10) from different random starting points to see if the sampled posterior distributions for each chain are the same.  Unfortunately, when the distributions all seem to be this same, this doesn't <span class="emphasis"><em>prove</em></span> that they have all converged to the equilibrium distribution.  However, if the distributions are different then you can reject either convergence or good mixing.</p>

    <div class="section"><div class="titlepage"><div><div><h3 class="title"><a name="idp58439968"></a>10.1. Definition of Convergence</h3></div></div></div>
      

      <p>Convergence refers to the the tendency of a Markov chain to
	to "forget" its starting value and become typical of its
	equilibrium distribution. Note that convergence is a property
	of the Markov chain itself, not of individual runs of the
	Markov chain.  Ideally a number of individual runs should be
	examined in order to determine how many initial iterations to
	discard as "burnin".
      </p>
    </div>
    
    <div class="section"><div class="titlepage"><div><div><h3 class="title"><a name="idp58442256"></a>10.2. Definition of Mixing</h3></div></div></div>
      
      <p>
	In MCMC, each sample is not fully independent of previous
	samples.  In fact, even after a Markov chain has converged,
	it can get "stuck" in one part of the parameter space for a
	long time, before jumping to an equally important part.  When
	this happens, each new sample contributes very little new
	information, and we need to obtain many more samples to get
	good precision on our parameter estimates.  In such a case, we say 
	that the chain isn't "mixing" well. 
      </p>
    </div>

    <div class="section"><div class="titlepage"><div><div><h3 class="title"><a name="idp58444272"></a>10.3. Diagnostics and multiple independent chains</h3></div></div></div>
      
    <p>
      Many of the following diagnostic measures require that you run the MCMC a number of
      different times.  In the text that follows, we refer to the
      different parameter files
      as <code class="filename">C1.p</code>, <code class="filename">C2.p</code>, 
      ..., <code class="filename">C<em class="replaceable"><code>n</code></em>.p</code> and the
      different tree samples as <code class="filename">C1.trees</code>, <code class="filename">C2.trees</code>,
      ..., <code class="filename">C<em class="replaceable"><code>n</code></em>.trees</code>.
      Substitute the actual names of the files.
    </p>
    </div>

    
    <div class="section"><div class="titlepage"><div><div><h3 class="title"><a name="idp58449872"></a>10.4. Diagnostics: Variation in split frequencies across runs (ASDSF/MSDSF)</h3></div></div></div>
      
      <div class="section"><div class="titlepage"><div><div><h4 class="title"><a name="idp58450128"></a>10.4.1. ASDSF and MSDSF</h4></div></div></div>
	
<p>
To calculate the ASDSF and MSDSF run:
</p><pre class="screen"><code class="prompt">%</code> trees-bootstrap C1.trees C2.trees ... C<em class="replaceable"><code>n</code></em>.trees &gt; partitions.bs</pre><p>	
For each split, the SDSF value is just the standard deviation across
runs of the Posterior Probabilities for that split.  By averaging the
resulting SDSF values across splits, we may obtain the ASDSF value
(Huelsenbeck and Ronquist 2001).  This is commonly considered
acceptable if it is &lt; 0.01.
</p>

<p>However, it is also useful to consider the maximum of the SDSF
  values (MSDSF).  This represents the range of variation in PP across
  the runs for the split with the most variation.
</p>
      </div>
      <div class="section"><div class="titlepage"><div><div><h4 class="title"><a name="idp58454864"></a>10.4.2. Split-frequency comparison plot</h4></div></div></div>
	
	<p>To generate the split-frequency comparison plot, you must have R installed.  Locate the script <code class="filename">compare-runs.R</code>.  Then run:
</p><pre class="screen"><code class="prompt">%</code> trees-bootstrap C1.trees C2.trees ... C<em class="replaceable"><code>n</code></em>.trees --LOD-table=LOD-table &gt; partitions.bs 
<code class="prompt">%</code> R --slave --vanilla --args LOD-table compare-SF.pdf &lt; compare-runs.R</pre><p>
	  Following <a class="ulink" href="http://dx.doi.org/10.1080/10635150600812544" target="_top">Beiko et al (2006)</a>, this displays the variation in
	  estimates of split frequencies across runs.  Splits are
	  arranged on the x-axis in increasing order of 
	  Posterior Probability (PP), which is obtained by averaging over
	  runs.  We then plot a vertical bar from the minimum PP to the
	  maximum PP.
	</p>
	</div>
    </div>


    <div class="section"><div class="titlepage"><div><div><h3 class="title"><a name="idp58459648"></a>10.5. Diagnostics: Potential Scale Reduction Factors (PSRF)</h3></div></div></div>
      
<p>
Potential Scale Redution Factors check that different runs have
similar posterior distributions.  Only numerical variables may have a
PSRF. To calculate the PSRF for each
numerical parameter, you may run: 

</p><pre class="screen"><code class="prompt">%</code> statreport C1.p C2.p ... C<em class="replaceable"><code>n</code></em>.p &gt; Report </pre><p>
The PSRF is a ratio of the width of the pooled distribution to the
average width of each distribution, and should ideally be 1.  The PSRF
is customarily considered to be small enough if it is less than 1.01.
</p>

<p>
We compare the PSRF based on the length of 80% credible intervals
(Brooks and Gelman 1998) and report the result as PSRF-80%CI.  For
integer-valued parameters, we avoid excessively large PSRF values by
subtracting 1 from the width of the pooled CI.
</p>

<p>
We also report a new PSRF that is more sensitive for integer
distributions.  For each individual distribution, we find the 80%
credible interval.  We divide the probability of that interval (which
may be more than 80%) by the probability of the same interval under the
pooled distribution.  The average of this measure over all
distributions gives us a PSRF that we report as PSRF-RCF.
</p>

<p>This convergence diagnostic gives a criterion for
detecting when a parameter value has stabilized at different
values in several independent runs, indicating a lack of
convergence. This situation might occur if different runs of
the Markov chain were trapped in different modes and failed to
adequately mix between modes.</p>
    </div>

    <div class="section"><div class="titlepage"><div><div><h3 class="title"><a name="idp58465168"></a>10.6. Diagnostics: Effective sample sizes (ESS)</h3></div></div></div>
      
      <div class="section"><div class="titlepage"><div><div><h4 class="title"><a name="idp58465488"></a>10.6.1. ESS for numerical values</h4></div></div></div>
	
      <p>To calculate the split ESS values, run:
</p><pre class="screen"><code class="prompt">%</code> statreport C1.p C2.p ... C<em class="replaceable"><code>n</code></em>.p &gt; Report </pre><p>
      We calculate effective sample sizes based on integrated
      autocorrelation times.  This method has the nice property that
      simply duplicating every sample does not increase the ESS.
      </p>

      <p>The
      program <a class="ulink" href="http://evolve.zoo.ox.ac.uk/software/tracer/" target="_top">Tracer</a>
      also computes ESS values.</p>
      </div>

      <div class="section"><div class="titlepage"><div><div><h4 class="title"><a name="idp58470560"></a>10.6.2. ESS for split frequencies</h4></div></div></div>
	
      <p>To calculate the split ESS values, run:
</p><pre class="screen"><code class="prompt">%</code> trees-bootstrap C1.trees C2.trees ... C<em class="replaceable"><code>n</code></em>.trees &gt; partitions.bs</pre><p>
      To compute the ESS for a split, we consider the presence or absence
      of a split in each iteration as a series of binary values.  We
      compute the integrated autocorrelation time for this binary
      sequence, which leads to an ESS.  This approach is similar to
      dividing the iterations into blocks and computing the ESS on the
      PP estimates in the blocks.  It is also similar to estimating
      the variance reduction under a block bootstrap.
      </p>
    </div>
</div>

    <div class="section"><div class="titlepage"><div><div><h3 class="title"><a name="idp58474160"></a>10.7. Diagnostics: Stabilization</h3></div></div></div>
      
      <div class="section"><div class="titlepage"><div><div><h4 class="title"><a name="idp58474416"></a>10.7.1. Stabilization of numerical values</h4></div></div></div>
	
<p>To obtain estimates of the stabilization time for each
numerical  parameter, you may run:
</p><pre class="screen"><code class="prompt">%</code> statreport C1.p &gt; Report </pre><p>
Each series of values is counted as having stabilized after
the series crosses its upper and then lower 95% confidence bounds
twice (if the initial value is below the median) or crosses its lower
and then upper confidence bounds twice (if the initial value is above
the median). The confidence bounds are those based on its
equilibrium distribution as calculated from the last third of the
values in the sequence.</p>
      </div>

      <div class="section"><div class="titlepage"><div><div><h4 class="title"><a name="idp58477856"></a>10.7.2. Stabilization of tree topologies and tree distances</h4></div></div></div>
	
	<p>In addition to examining convergence diagnostics for continuous
	parameters, it is important to examine convergence diagnostics
	for the topology as well
	(<a class="ulink" href="http://dx.doi.org/10.1080/10635150600812544" target="_top">Beiko
	et al 2006</a>).  In theory, we recommend the web tool <a class="ulink" href="http://ceb.csit.fsu.edu/awty/" target="_top">Are We There Yet (AWTY)</a> (Wilgenbush et al, 2004).  However, AWTY gives incorrect results if you upload plain NEWICK tree samples -- which is what BAli-Phy outputs.  Therefore, if you wish to use AWTY, you must convert the tree samples files to NEXUS before you upload them to AWTY in order to get correct results.</p>

<p>It is also be possible to assess stabilization of tree topologies using tools distributed with <span class="application">bali-phy</span> by using commands like the following.  Here, sub-sampling and burnin does not apply to the equilibrium tree files. Also, note that you need to manually construct the equilibrium samples, which we recommend to contain at least 500 trees; you might do this by sub-sampling using the <span class="application">BAli-Phy</span> tool <span class="command"><strong>sub-sample</strong></span>.</p>

<div class="orderedlist"><ol class="orderedlist" type="1"><li class="listitem"><p>To report the average distances within and between two tree samples:
</p><pre class="screen"><code class="prompt">%</code> trees-distances --skip=<em class="replaceable"><code>burnin</code></em> --sub-sample=<em class="replaceable"><code>factor</code></em> compare <code class="filename">C1.trees</code> <code class="filename">C2.trees</code></pre><p>
</p></li><li class="listitem"><p>To compute the distance from each tree in C1.trees to all trees equilibrium.trees, as a time series:
</p><pre class="screen"><code class="prompt">%</code> trees-distances --skip=<em class="replaceable"><code>burnin</code></em> --sub-sample=<em class="replaceable"><code>factor</code></em> convergence <code class="filename">C1.trees</code> <code class="filename">equilibrium.trees</code></pre><p>
</p></li><li class="listitem"><p>To assess when the above time series stabilizes:
</p><pre class="screen"><code class="prompt">%</code> trees-distances --skip=<em class="replaceable"><code>burnin</code></em> --sub-sample=<em class="replaceable"><code>factor</code></em> converged <code class="filename">C1.trees</code> <code class="filename">equilibrium.trees</code></pre><p>
The stabilization criterion is the same one described above for numerical values.
</p></li></ol></div>

<p>Note that the running time is the product of the number of trees in the two files.  Therefore, comparing two complete tree samples without sub-sampling will take too long.</p>

   </div>
	
    </div>

    
  </div>


  <div class="section"><div class="titlepage"><div><div><h2 class="title" style="clear: both"><a name="auxiliary_tools"></a>11. Auxiliary tools</h2></div></div></div>
    

    <p>Most of these tools will describe their options if given the "<strong class="userinput"><code>--help</code></strong>" argument on the command line.</p>

    <div class="section"><div class="titlepage"><div><div><h3 class="title"><a name="idp58497504"></a>11.1. alignment-find</h3></div></div></div>
      
      <p>Usage: alignment-find [OPTIONS] &lt; <em class="replaceable"><code>alignments-file</code></em></p>
      <p>Find the last (or first) FastA alignment in a file.</p>
    </div>

    <div class="section"><div class="titlepage"><div><div><h3 class="title"><a name="idp58499664"></a>11.2. alignment-draw</h3></div></div></div>
      

      <p>alignment-draw <em class="replaceable"><code>alignment-file</code></em> [<em class="replaceable"><code>AU-file</code></em>] [OPTIONS]</p>
      <p>Draw an alignment to HTML, optionally coloring residues by AU.</p>
    </div>

    <div class="section"><div class="titlepage"><div><div><h3 class="title"><a name="idp58502736"></a>11.3. alignment-thin</h3></div></div></div>
      
      <p>alignment-thin <em class="replaceable"><code>alignment-file</code></em> <em class="replaceable"><code>tree-file</code></em> [OPTIONS]</p>
      <p>Remove taxa from an alignment to preserve the most
	sequence diversity, as measured by the  total length of the tree
	for the remaining taxa.</p> 
    </div>

    <div class="section"><div class="titlepage"><div><div><h3 class="title"><a name="idp58506000"></a>11.4. alignment-chop-internal</h3></div></div></div>
      
      <p>alignment-chop-internal <em class="replaceable"><code>alignment-file</code></em> [OPTIONS]</p>
      <p>Remove ancestral sequences from an alignment.  (This
	probably only works for alignments output by bali-phy.) </p>
    </div>

    <div class="section"><div class="titlepage"><div><div><h3 class="title"><a name="idp58508464"></a>11.5. alignment-info</h3></div></div></div>
      
      <p>alignment-info <em class="replaceable"><code>alignment-file</code></em> <em class="replaceable"><code>tree-file</code></em> [OPTIONS]</p>
      <p>Display basic information about the alignment, including
	its length, the number of sequences, columns that are constant
	or informative, letter frequencies, etc.</p>
      <p>If a tree is supplied, then the unweighted parsimony score
	is given as well.</p>
    </div>
    
    <div class="section"><div class="titlepage"><div><div><h3 class="title"><a name="idp58512048"></a>11.6. alignment-indices</h3></div></div></div>
      
      <p>alignment-indices <em class="replaceable"><code>alignment-file</code></em> [OPTIONS]</p>
      <p>Show the alignment in terms of the index of each character
	in its sequence.  Each line in this file corresponds to one
	alignment column.  This can be useful in producing alignment
	constraint files.</p>
      <p>Also, you can specify which columns to keep using the
	--columns option.
      </p>
    </div>
    
    <div class="section"><div class="titlepage"><div><div><h3 class="title"><a name="idp58515312"></a>11.7. alignment-cat</h3></div></div></div>
      
      <p>alignment-cat <em class="replaceable"><code>file1</code></em>
	[<em class="replaceable"><code>file2</code></em> ...]</p>
      <p>Concatenate several alignments (with the same sequence
	names) end-to-end.</p>
    </div>
    
    <div class="section"><div class="titlepage"><div><div><h3 class="title"><a name="idp58518352"></a>11.8. trees-consensus</h3></div></div></div>
      
      <p>Usage: trees-consensus <em class="replaceable"><code>file</code></em>
	[OPTIONS] </p>
      <p>This program analyzes the tree sample contained in
	<em class="replaceable"><code>file</code></em>.  It reports the MAP topology, the
	supported taxa partitions (including partial partitions), and the
	majority consensus topology.
      </p> 
    </div>

    <div class="section"><div class="titlepage"><div><div><h3 class="title"><a name="idp58521728"></a>11.9. trees-bootstrap</h3></div></div></div>
      
      <p>Usage: trees-bootstrap <em class="replaceable"><code>file1</code></em>
	[<em class="replaceable"><code>file2</code></em> ... ] --predicates
	<em class="replaceable"><code>predicate-file</code></em> [OPTIONS] </p>
      <p>This program analyzes the tree samples contained in
	<em class="replaceable"><code>file1</code></em>, <em class="replaceable"><code>file2</code></em>,
	etc.  It gives the support of each tree sample for each predicate in
	<em class="replaceable"><code>predicate-file</code></em>, and reports a confidence
	interval based on the block bootstrap.
      </p> 

      <p>Each predicate is the intersection of a set of partitions, and
	is specified as a list of partitions or (multifurcating) trees, one
	per line.  Predicates are separated by blank lines.
      </p>
    </div>

    <div class="section"><div class="titlepage"><div><div><h3 class="title"><a name="idp58527088"></a>11.10. trees-to-SRQ</h3></div></div></div>
      

      <p>Usage: trees-to-SRQ <em class="replaceable"><code>predicate-file</code></em> [OPTIONS] <em class="replaceable"><code>trees-file</code></em> </p>

      <p>This program analyzes the tree samples contained in
	<em class="replaceable"><code>trees-file</code></em>.  It uses them to produce an
	SRQ plot for each predicate in
	<em class="replaceable"><code>predicate-file</code></em>.  Plots are produced in
	<span class="application">gnuplot</span> format, with one point per line
	and with plots separated by a blank line.</p>

      <p>If <strong class="userinput"><code>--mode sum </code></strong> is specified, then a "sum"
	plot is produced instead of an SRQ plot.  In this plot, the slope of
	the curve corresponds to the posterior probability of the event.  If the
	<strong class="userinput"><code>--invert</code></strong> option is used then the slope of the
	curve correspond to the probability of the inverse event.  This is
	recommended if the probability of the event is near 1.0, because the
	sum plot does not distinguish variation in probabilities near 1.0 well.
      </p>

    </div>

  </div>

  <div class="section"><div class="titlepage"><div><div><h2 class="title" style="clear: both"><a name="FAQ"></a>12. Frequently Asked Questions (FAQ)</h2></div></div></div>
    
    <div class="section"><div class="titlepage"><div><div><h3 class="title"><a name="idp58534000"></a>12.1. Input files</h3></div></div></div>
      <div class="qandaset"><a name="idp58535136"></a><dl><dt>12.1.1. <a href="#idp58536080">Does BAli-Phy accept the wildcard characters "N" or "X"?  How does it treat them?</a></dt><dt>12.1.2. <a href="#idp58538864">Does BAli-Phy accept "?" characters?</a></dt><dt>12.1.3. <a href="#idp58542560">Does BAli-Phy accept the characters "R" and "Y", etc.?</a></dt></dl><table border="0" style="width: 100%;"><colgroup><col align="left" width="1%"><col></colgroup><tbody>
	  <tr class="question"><td align="left" valign="top"><a name="idp58536080"></a><a name="idp58536336"></a><p><b>12.1.1.</b></p></td><td align="left" valign="top"><p>Does BAli-Phy accept the wildcard characters "N" or "X"?  How does it treat them?</p></td></tr>
	  <tr class="answer"><td align="left" valign="top"></td><td align="left" valign="top"><p>Yes, BAli-Phy accepts the wildcard characters "N"
	    (for DNA) and "X" (for proteins).  These characters
	    indicate that some letter is present (as opposed to a
	    gap), but that you don't know <span class="emphasis"><em>which</em></span>
	    letter it is.  
	    </p></td></tr>
	
	  <tr class="question"><td align="left" valign="top"><a name="idp58538864"></a><a name="idp58537808"></a><p><b>12.1.2.</b></p></td><td align="left" valign="top"><p>Does BAli-Phy accept "?" characters?</p></td></tr>
	  <tr class="answer"><td align="left" valign="top"></td><td align="left" valign="top"><p>
	      No.  "?" characters are often used to indicate
	      <span class="emphasis"><em>either</em></span> letter presence (e.g. "N",
	      "X") <span class="emphasis"><em>or</em></span> absence (e.g. "-").
	      BAli-phy will insist that you replace each "?" with
	      either "N"/"X" or "-" to indicate which one you mean. 
	    </p><p>(Most programs ignore indels and consider only
	    substitutions, and in that case "N" and "-" have the same
	    effect on the likelihood or parsimony score.  However,
	    since BAli-Phy takes indels into account, these two
	    alternatives are quite different.)
	    </p></td></tr>
	
	<tr class="question"><td align="left" valign="top"><a name="idp58542560"></a><a name="idp58542816"></a><p><b>12.1.3.</b></p></td><td align="left" valign="top"><p>Does BAli-Phy accept the characters "R" and "Y", etc.?</p></td></tr>
	  <tr class="answer"><td align="left" valign="top"></td><td align="left" valign="top"><p>
	      Yes.  BAli-Phy accepts the characters "Y", "R", "W", and "S" for
	      DNA, RNA, and Codon alphabets.  However, it does not
	      accept the characters "K", "M", "B", "D", "H",  and "V".
	      BAli-Phy also accepts the characters "B", "Z", and "J" 
	      for amino acids.  These characters indicate partial
	      knowledge about a letter.  For example, "R" indicates
	      that a nucleotide is present, and is a puRine ("A" or
	      "G"). "J" indicates that an amino acid is present and is
	      either "I" or "L".  
	    </p><p>
	      (Note that sequences sometimes contain such ambiguity
	      codes because the DNA that was sequenced contains
	      <span class="emphasis"><em>both</em></span> values.  This might occur when
	      sequencing a heterozygote or when sequencing pooled DNA
	      from several individuals.  However, the model in
	      BAli-Phy (and other phylogeny inference programs) is
	      that only one letter is correct, but we do not know
	      which one it is.  This is probably not problematic when
	      dealing with pooled sequences, but should be considered.)
	    </p></td></tr>
	</tbody></table></div>
    </div>

    <div class="section"><div class="titlepage"><div><div><h3 class="title"><a name="idp58545728"></a>12.2. Running <span class="command"><strong>bali-phy</strong></span>.</h3></div></div></div>

      <div class="qandaset"><a name="idp58548480"></a><dl><dt>12.2.1. <a href="#idp58549440">Can I fix the alignment and ignore indel information, like MrBayes and other MCMC programs?</a></dt><dt>12.2.2. <a href="#idp58552784">Can I fix the tree topology, while allowing the alignment to vary?</a></dt><dt>12.2.3. <a href="#idp58555664">Can I fix the tree topology and relative branch lengths, while allowing the alignment to vary?</a></dt><dt>12.2.4. <a href="#idp58558768">Can I fix the tree topology and absolute branch lengths in all data partitions, while allowing the alignment to vary?</a></dt></dl><table border="0" style="width: 100%;"><colgroup><col align="left" width="1%"><col></colgroup><tbody>
	  <tr class="question"><td align="left" valign="top"><a name="idp58549440"></a><a name="idp58549696"></a><p><b>12.2.1.</b></p></td><td align="left" valign="top"><p>Can I fix the alignment and ignore indel information, like MrBayes and other MCMC programs?</p></td></tr>
	  <tr class="answer"><td align="left" valign="top"></td><td align="left" valign="top"><p>Yes.  Add <strong class="userinput"><code>-t</code></strong> or <strong class="userinput"><code>--traditional</code></strong> on the command line.  This has the same effect as <strong class="userinput"><code>--imodel none</code></strong>.</p></td></tr>
	
	  <tr class="question"><td align="left" valign="top"><a name="idp58552784"></a><a name="idp58553104"></a><p><b>12.2.2.</b></p></td><td align="left" valign="top"><p>Can I fix the tree topology, while allowing the alignment to vary?</p></td></tr>
	  <tr class="answer"><td align="left" valign="top"></td><td align="left" valign="top"><p>Yes.  Add <strong class="userinput"><code>--disable=topology --tree=<em class="replaceable"><code>treefile</code></em></code></strong> on the command line.</p></td></tr>
	
	  <tr class="question"><td align="left" valign="top"><a name="idp58555664"></a><a name="idp58555920"></a><p><b>12.2.3.</b></p></td><td align="left" valign="top"><p>Can I fix the tree topology and <span class="emphasis"><em>relative</em></span> branch lengths, while allowing the alignment to vary?</p></td></tr>
	  <tr class="answer"><td align="left" valign="top"></td><td align="left" valign="top"><p>Yes.  Add <strong class="userinput"><code>--disable=tree --tree=<em class="replaceable"><code>treefile</code></em></code></strong> on the command line.</p></td></tr>
	
	  <tr class="question"><td align="left" valign="top"><a name="idp58558768"></a><a name="idp58559024"></a><p><b>12.2.4.</b></p></td><td align="left" valign="top"><p>Can I fix the tree topology and <span class="emphasis"><em>absolute</em></span> branch lengths <span class="emphasis"><em>in all data partitions</em></span>, while allowing the alignment to vary?</p></td></tr>
	  <tr class="answer"><td align="left" valign="top"></td><td align="left" valign="top"><p>Yes.  Add <strong class="userinput"><code>--disable=tree --tree=<em class="replaceable"><code>treefile</code></em> --fix=mu1 ... --fix-mu<em class="replaceable"><code>n</code></em></code></strong> on the command line.</p></td></tr>
	</tbody></table></div>

    </div>

    <div class="section"><div class="titlepage"><div><div><h3 class="title"><a name="idp58562960"></a>12.3. Run-time error messages</h3></div></div></div>
      
      
      <div class="qandaset"><a name="idp58563296"></a><dl><dt>12.3.1. <a href="#idp58564544">I tried to use --smodel LG+gamma[6] and I got an error message "bali-phy: No match."  What gives?</a></dt></dl><table border="0" style="width: 100%;"><colgroup><col align="left" width="1%"><col></colgroup><tbody>
	  <tr class="question"><td align="left" valign="top"><a name="idp58564544"></a><a name="idp58564800"></a><p><b>12.3.1.</b></p></td><td align="left" valign="top"><p>I tried to use <strong class="userinput"><code>--smodel LG+gamma[6]</code></strong> and I got an error message "bali-phy: No match."  What gives?</p></td></tr>
	  <tr class="answer"><td align="left" valign="top"></td><td align="left" valign="top"><p>You are probably using the C-shell as your command line shell.  It is trying to interpret <strong class="userinput"><code>LG+gamma[6]</code></strong> as an array before running the command, and it is not succeeding.  Therefore, it doesn't even run <span class="command"><strong>bali-phy</strong></span>.</p><p>To avoid this, put a backslash in front of the first "[" and write <strong class="userinput"><code>--smodel LG+gamma\[6]</code></strong>.  This will keep the C-shell from interfering with your command.
	    </p></td></tr>
	</tbody></table></div>
    </div>

    <div class="section"><div class="titlepage"><div><div><h3 class="title"><a name="idp58568992"></a>12.4. Stopping <span class="command"><strong>bali-phy</strong></span>.</h3></div></div></div>
      

      <div class="qandaset"><a name="idp58571376"></a><dl><dt>12.4.1. <a href="#idp58571728">Why is bali-phy still
	      running? How long will it take?</a></dt><dt>12.4.2. <a href="#idp58574064">How do I stop a bali-phy
	  run on my personal computer?</a></dt><dt>12.4.3. <a href="#idp58582048">How do I stop a bali-phy
	  run on a computing cluster?</a></dt><dt>12.4.4. <a href="#idp58586224">So, how can I know when to stop it?</a></dt><dt>12.4.5. <a href="#idp58588256">How can I tell when the chain has converged?</a></dt><dt>12.4.6. <a href="#idp58590720">How can I check how many iterations the chain
	      has finished?</a></dt></dl><table border="0" style="width: 100%;"><colgroup><col align="left" width="1%"><col></colgroup><tbody>
	  <tr class="question"><td align="left" valign="top"><a name="idp58571728"></a><a name="idp58571984"></a><p><b>12.4.1.</b></p></td><td align="left" valign="top"><p>Why is <span class="command"><strong>bali-phy</strong></span> still
	      running? How long will it take?</p></td></tr>
	  <tr class="answer"><td align="left" valign="top"></td><td align="left" valign="top"><p>It runs until you stop it.  Stop it when its
	      done.</p></td></tr>
	
	  <tr class="question"><td align="left" valign="top"><a name="idp58574064"></a><a name="idp58574320"></a><p><b>12.4.2.</b></p></td><td align="left" valign="top"><p>How do I stop a <span class="command"><strong>bali-phy</strong></span>
	  run on my personal computer?</p></td></tr>
	  <tr class="answer"><td align="left" valign="top"></td><td align="left" valign="top"><p>Simply kill the process -- there is no special
	    command to stop <span class="command"><strong>bali-phy</strong></span>. If you are
	    running it on your personal workstation, then you can use
	    the command <span class="command"><strong>kill</strong></span>.  To do that, you need
	    to find the PID (process ID) of the running program.  You
	    can find this by examining the beginning of the file
	    <code class="filename">C1.out</code>.  For 
	    example:
</p><pre class="screen"><code class="prompt">%</code> less 5d-1/C1.out
command: bali-phy 5d.fasta
start time: Thu Oct  3 15:18:17 2013

VERSION: 2.2.0  [2.2 commit 66ad8405+]  (Oct 03 2013 14:47:48)
BUILD: Oct  3 2013 15:13:10
ARCH: x86_64-unknown-linux-gnu
COMPILER: GCC 4.8.1
FLAGS: -isystem $(top_srcdir)/boost/include -ffast-math -DNDEBUG -DNDEBUG_DP -funroll-loops -std=c++11 -finline-limit=1000   -pipe -O3
directory: /home/bredelings/Devel/bali-phy/Runs
subdirectory: 5d-1
hostname: mutant
<span class="emphasis"><em>PID: 21415</em></span>

random seed = 10968853200429096211

...
</pre><p>
Here the PID is 21415.  Therefore you can type:
</p><pre class="screen"><code class="prompt">%</code> kill 21415</pre><p>
On some operating systems you can also type:
</p><pre class="screen"><code class="prompt">%</code> killall bali-phy</pre><p>
However, be aware that this will terminate <span class="emphasis"><em>all</em></span> of
your <span class="command"><strong>bali-phy</strong></span> runs on that computer.
	    </p></td></tr>
	
	  <tr class="question"><td align="left" valign="top"><a name="idp58582048"></a><a name="idp58582304"></a><p><b>12.4.3.</b></p></td><td align="left" valign="top"><p>How do I stop a <span class="command"><strong>bali-phy</strong></span>
	  run on a computing cluster?</p></td></tr>
	  <tr class="answer"><td align="left" valign="top"></td><td align="left" valign="top"><p>Simply terminate the submitted job.  The specific command
	    to terminate a job will depend on the queue manager that
	    is installed on your cluster.  Examine the documentation
	    for your cluster, or ask your cluster support staff how to delete
	    running jobs on your cluster.
	    </p><p>As an example, if the SGE software is used
	    to submit jobs, then the command <span class="command"><strong>qstat</strong></span>
	    should list your jobs and their job ID numbers (which is
	    different than the process ID number).  You can then use
	    the command <span class="command"><strong>qdel</strong></span> to delete jobs by ID
	    number.  The SGE documentation describes how to use these
	    commands. 
	    </p></td></tr>
	
	  <tr class="question"><td align="left" valign="top"><a name="idp58586224"></a><a name="idp58586544"></a><p><b>12.4.4.</b></p></td><td align="left" valign="top"><p>So, how can I know when to stop it?</p></td></tr>
	  <tr class="answer"><td align="left" valign="top"></td><td align="left" valign="top"><p>You can stop when it has both converged and also run for long enough to give
	      you &gt;1000 effectively independent samples.  </p></td></tr>
	
	  <tr class="question"><td align="left" valign="top"><a name="idp58588256"></a><a name="idp58588512"></a><p><b>12.4.5.</b></p></td><td align="left" valign="top"><p>How can I tell when the chain has converged?</p></td></tr>
	  <tr class="answer"><td align="left" valign="top"></td><td align="left" valign="top"><p>See section <a class="xref" href="#mixing_and_convergence" title="10. Convergence and Mixing: Is it done yet?">Section 10, &#8220;Convergence and Mixing: Is it done yet?&#8221;</a>.</p></td></tr>
	
	  <tr class="question"><td align="left" valign="top"><a name="idp58590720"></a><a name="idp58591040"></a><p><b>12.4.6.</b></p></td><td align="left" valign="top"><p>How can I check how many iterations the chain
	      has finished?</p></td></tr>
	  <tr class="answer"><td align="left" valign="top"></td><td align="left" valign="top"><p>Run <span class="command"><strong>wc -l C1.p</strong></span> inside the output
	      directory, and subtract 2.
	    </p></td></tr>
	</tbody></table></div>
    </div>

    <div class="section"><div class="titlepage"><div><div><h3 class="title"><a name="idp58593248"></a>12.5. Interpreting the results.</h3></div></div></div>
      

      <div class="qandaset"><a name="idp58593504"></a><dl><dt>12.5.1. <a href="#idp58594800">How do I compute the clade support?</a></dt><dt>12.5.2. <a href="#idp58596864">How do I compute the split/bi-partition support?</a></dt></dl><table border="0" style="width: 100%;"><colgroup><col align="left" width="1%"><col></colgroup><tbody>
	  <tr class="question"><td align="left" valign="top"><a name="idp58594800"></a><a name="idp58595056"></a><p><b>12.5.1.</b></p></td><td align="left" valign="top"><p>How do I compute the clade support?</p></td></tr>
	  <tr class="answer"><td align="left" valign="top"></td><td align="left" valign="top"><p>Actually, BAli-Phy uses unrooted trees, so it only estimates bi-partition support.  A bi-partition is a division of taxa into two groups, but it does not specify which group contains the root. </p></td></tr>
	
	  <tr class="question"><td align="left" valign="top"><a name="idp58596864"></a><a name="idp58597184"></a><p><b>12.5.2.</b></p></td><td align="left" valign="top"><p>How do I compute the split/bi-partition support?</p></td></tr>
	  <tr class="answer"><td align="left" valign="top"></td><td align="left" valign="top"><p>After you analyze the output (<a class="xref" href="#analysis" title="6.5. Summarizing the output - scripted">Section 6.5, &#8220;Summarizing the output - scripted&#8221;</a>), the partition support is indicated in
	      <code class="filename">Results/consensus</code> and in <code class="filename">Results/c50.PP.tree</code>. </p></td></tr>
	</tbody></table></div>
    </div>

    <div class="section"><div class="titlepage"><div><div><h3 class="title"><a name="idp58600656"></a>12.6. How do I...</h3></div></div></div>
      
      <div class="qandaset"><a name="idp58600912"></a><dl><dt>12.6.1. <a href="#idp58602064">How do I concatenate alignments?</a></dt><dt>12.6.2. <a href="#idp58605712">How do I select columns from an alignment?</a></dt><dt>12.6.3. <a href="#idp58608672">How do I create an alignment-constraint file
	      from an alignment?</a></dt></dl><table border="0" style="width: 100%;"><colgroup><col align="left" width="1%"><col></colgroup><tbody>
	  <tr class="question"><td align="left" valign="top"><a name="idp58602064"></a><a name="idp58602320"></a><p><b>12.6.1.</b></p></td><td align="left" valign="top"><p>How do I concatenate alignments?</p></td></tr>
	  <tr class="answer"><td align="left" valign="top"></td><td align="left" valign="top"><p>
 	    </p><pre class="screen"><code class="prompt">%</code> alignment-cat <em class="replaceable"><code>filename1.fasta</code></em> <em class="replaceable"><code>filename2.fasta</code></em> &gt; result.fasta</pre><p>
	      The alignments must have the same sequence names, but
	      the names need not be in the same order.
	    </p></td></tr>
	
	  <tr class="question"><td align="left" valign="top"><a name="idp58605712"></a><a name="idp58605968"></a><p><b>12.6.2.</b></p></td><td align="left" valign="top"><p>How do I select columns from an alignment?</p></td></tr>
	  <tr class="answer"><td align="left" valign="top"></td><td align="left" valign="top"><p>
	    </p><pre class="screen"><code class="prompt">%</code> alignment-cat -c1-10,50-100,600- <em class="replaceable"><code>filename.fasta</code></em> &gt; result.fasta</pre><p>
	    The resulting alignment will contain the selected columns
	    in the order you specified.
	  </p></td></tr>
	
	  <tr class="question"><td align="left" valign="top"><a name="idp58608672"></a><a name="generating_constraint_files"></a><p><b>12.6.3.</b></p></td><td align="left" valign="top"><p>How do I create an alignment-constraint file
	      from an alignment?</p></td></tr>
	  <tr class="answer"><td align="left" valign="top"></td><td align="left" valign="top"><p>To constrain the alignment to match some alignment
	      file <em class="replaceable"><code>filename.fasta</code></em> in columns
	      100, 200-250, and 300, run:
	      </p><pre class="screen"><code class="prompt">%</code> alignment-indices -c100,200-250,300 <em class="replaceable"><code>filename.fasta</code></em> &gt; filename.constraint</pre><p>
	    </p></td></tr>	
	</tbody></table></div>
    </div>
  </div>

</div></body></html>
