<html><head><meta http-equiv="Content-Type" content="text/html; charset=ISO-8859-1"><title>BAli-Phy Tutorial</title><link rel="stylesheet" type="text/css" href="docbook.css"><meta name="generator" content="DocBook XSL Stylesheets V1.78.1"></head><body bgcolor="white" text="black" link="#0000FF" vlink="#840084" alink="#0000FF"><div lang="en" class="article"><div class="titlepage"><div><div><h2 class="title"><a name="idp53012672"></a><span class="application">BAli-Phy</span> Tutorial</h2></div><div><div class="author"><h3 class="author"><span class="firstname">Benjamin</span> <span class="surname">Redelings</span></h3></div></div></div><hr></div><div class="toc"><p><b>Table of Contents</b></p><dl class="toc"><dt><span class="section"><a href="#intro">1. Introduction</a></span></dt><dt><span class="section"><a href="#work_directory">2. Setting up the <code class="filename">~/Work</code> directory</a></span></dt><dt><span class="section"><a href="#command_line_options">3. Command line options</a></span></dt><dd><dl><dt><span class="section"><a href="#idp53035936">3.1. RNA</a></span></dt><dt><span class="section"><a href="#idp53044720">3.2. Amino Acids</a></span></dt><dt><span class="section"><a href="#idp53047168">3.3. Codons</a></span></dt></dl></dd><dt><span class="section"><a href="#idp53052016">4. Output</a></span></dt><dd><dl><dt><span class="section"><a href="#idp53057648">4.1. Inspecting output files</a></span></dt><dt><span class="section"><a href="#idp53068560">4.2. Summarizing the output</a></span></dt></dl></dd><dt><span class="section"><a href="#idp53085376">5. Multi-gene analyses</a></span></dt></dl></div>
  
    
  <div class="section"><div class="titlepage"><div><div><h2 class="title" style="clear: both"><a name="intro"></a>1. Introduction</h2></div></div></div>
<p>
This tutorial contains step-by-step commands, assuming that you have already installed bali-phy in <code class="filename">~/local/</code>.  The <a class="ulink" href="http://www.bali-phy.org/README.html" target="_top">manual</a> contains more detailed explanations for many things, so you might want to refer to it during the tutorial.</p>
  </div>
  <div class="section"><div class="titlepage"><div><div><h2 class="title" style="clear: both"><a name="work_directory"></a>2. Setting up the <code class="filename">~/Work</code> directory</h2></div></div></div>

<p>Go to your home directory:
</p><pre class="screen"><code class="prompt">%</code> <strong class="userinput"><code>cd ~</code></strong>
</pre><p>Make a directory called Work inside it:
</p><pre class="screen"><code class="prompt">%</code> <strong class="userinput"><code>mkdir Work</code></strong>
</pre><p>Go into the <code class="filename">Work</code> directory:
</p><pre class="screen"><code class="prompt">%</code> <strong class="userinput"><code>cd Work</code></strong>
</pre><p>Make a shortcut (a symbolic link) to the examples directory:
</p><pre class="screen"><code class="prompt">%</code> <strong class="userinput"><code>ln -s ~/local/share/bali-phy/examples examples</code></strong>
</pre><p>Take a look inside the <code class="filename">examples</code> directory:
</p><pre class="screen"><code class="prompt">%</code> <strong class="userinput"><code>ls examples</code></strong>
</pre><p>Take a look at an input file:
</p><pre class="screen"><code class="prompt">%</code> <strong class="userinput"><code>less examples/5S-rRNA/5d.fasta</code></strong>
</pre><p>Take a look at an input file:
</p><pre class="screen"><code class="prompt">%</code> <strong class="userinput"><code>alignment-info examples/5S-rRNA/5d.fasta</code></strong>
</pre>
  </div>

  <div class="section"><div class="titlepage"><div><div><h2 class="title" style="clear: both"><a name="command_line_options"></a>3. Command line options</h2></div></div></div>

<p>
What version of bali-phy are you running?  When was it compiled?  Which compiler?  For what computer type?
</p><pre class="screen"><code class="prompt">%</code> <strong class="userinput"><code>bali-phy -v</code></strong>
</pre><p>Look at the list of command line options:
</p><pre class="screen"><code class="prompt">%</code> <strong class="userinput"><code>bali-phy --help </code></strong>
</pre><p>Look at them with the ability to scroll back:
</p><pre class="screen"><code class="prompt">%</code> <strong class="userinput"><code>bali-phy --help | less</code></strong>
</pre><p>Some options have a short form which is a single letter:
</p><pre class="screen"><code class="prompt">%</code> <strong class="userinput"><code>bali-phy -h | less</code></strong>
</pre>

<div class="section"><div class="titlepage"><div><div><h3 class="title"><a name="idp53035936"></a>3.1. RNA</h3></div></div></div>
<p>
Analyze a data set, but don't begin MCMC.  (This is useful to know if the analysis works, what model will be used,
compute likelihoods, etc.)
</p><pre class="screen"><code class="prompt">%</code> <strong class="userinput"><code>bali-phy --test examples/5S-rRNA/5d.fasta</code></strong>
</pre><p>Finally, run an analysis!  (This is just 10 iterations, so its not a real run.)
</p><pre class="screen"><code class="prompt">%</code> <strong class="userinput"><code>bali-phy examples/5S-rRNA/5d.fasta --iterations=10</code></strong>
</pre><p>If you specify <em class="parameter"><code>--imodel=none</code></em>, then the alignment won't be estimated, and indels will be ignored (just like <span class="application">MrBayes</span>).
</p><pre class="screen"><code class="prompt">%</code> <strong class="userinput"><code>bali-phy examples/5S-rRNA/5d.fasta --iterations=10 --imodel=none</code></strong>
</pre><p>You can specify the alphabet, substitution model, insertion/deletion model, etc.
Defaults are used if you don't specify.
</p><pre class="screen"><code class="prompt">%</code> <strong class="userinput"><code>bali-phy examples/5S-rRNA/5d.fasta --iterations=10 --alphabet=DNA --smodel=TN --imodel=RS07 </code></strong>
</pre><p>You can change this to the GTR, if you want:
</p><pre class="screen"><code class="prompt">%</code> <strong class="userinput"><code>bali-phy examples/5S-rRNA/5d.fasta --iterations=10 --alphabet=DNA --smodel=GTR --imodel=RS07 </code></strong>
</pre><p>You can add gamma[4]+INV rate heterogeneity:
</p><pre class="screen"><code class="prompt">%</code> <strong class="userinput"><code>bali-phy examples/5S-rRNA/5d.fasta --iterations=10 --alphabet=DNA --smodel=GTR+gamma[4]+INV --imodel=RS07</code></strong>
</pre>
</div>

<div class="section"><div class="titlepage"><div><div><h3 class="title"><a name="idp53044720"></a>3.2. Amino Acids</h3></div></div></div>
<p>
When the data set contains amino acids, the default is amino acids:
</p><pre class="screen"><code class="prompt">%</code> <strong class="userinput"><code>bali-phy examples/EF-Tu/12d.fasta --iterations=10</code></strong>
</pre>
</div>

<div class="section"><div class="titlepage"><div><div><h3 class="title"><a name="idp53047168"></a>3.3. Codons</h3></div></div></div>
<p>
What alphabet is used here?  What substitution model?
</p><pre class="screen"><code class="prompt">%</code> <strong class="userinput"><code>bali-phy examples/HIV/chain-2005/env-clustal-codons.fasta --test</code></strong>
</pre><p>What happens when trying to use the M0 model (e.g. positive selection)?
</p><pre class="screen"><code class="prompt">%</code> <strong class="userinput"><code>bali-phy examples/HIV/chain-2005/env-clustal-codons.fasta --test --smodel=M0 </code></strong>
</pre><p>The M0 model requires a codon alphabet:
</p><pre class="screen"><code class="prompt">%</code> <strong class="userinput"><code>bali-phy examples/HIV/chain-2005/env-clustal-codons.fasta --test --smodel=M0 --alphabet=Codons </code></strong>
</pre>
</div>
  </div>


<div class="section"><div class="titlepage"><div><div><h2 class="title" style="clear: both"><a name="idp53052016"></a>4. Output</h2></div></div></div>
<p>
See <a class="ulink" href="http://www.bali-phy.org/README.html#output" target="_top">Section 6: Output</a> of the manual for more information about this section.
</p>
<p>
Try running an analysis with a few more iterations.  The &amp; makes the command run in the background, so that you can run other commands before it finishes.
</p><pre class="screen"><code class="prompt">%</code> <strong class="userinput"><code>bali-phy examples/5S-rRNA/25-muscle.fasta --iterations=200 &amp;</code></strong>
</pre><p>Run another copy of the same analysis:
</p><pre class="screen"><code class="prompt">%</code> <strong class="userinput"><code>bali-phy examples/5S-rRNA/25-muscle.fasta --iterations=200 &amp;</code></strong>
</pre><p>You can take a look at your running jobs:
</p><pre class="screen"><code class="prompt">%</code> <strong class="userinput"><code>jobs</code></strong>
</pre>
<div class="section"><div class="titlepage"><div><div><h3 class="title"><a name="idp53057648"></a>4.1. Inspecting output files</h3></div></div></div>
<p>
Look at the directories that were created to store the output files:
</p><pre class="screen"><code class="prompt">%</code> <strong class="userinput"><code>ls</code></strong>
<code class="prompt">%</code> <strong class="userinput"><code>ls 25-muscle-1/</code></strong>
<code class="prompt">%</code> <strong class="userinput"><code>ls 25-muscle-2/</code></strong>
</pre><p>See how many iterations have been completed so far:
</p><pre class="screen"><code class="prompt">%</code> <strong class="userinput"><code>wc -l 25-muscle-1/C1.p 25-muscle-2/C1.p</code></strong>
</pre><p>Wait a second, and repeat the command.
</p><pre class="screen"><code class="prompt">%</code> <strong class="userinput"><code>wc -l 25-muscle-1/C1.p 25-muscle-2/C1.p</code></strong>
</pre><p>See if you can determine the following information from the beginning of the C1.out file:
What command was run?
When was it run?
Which computer was it run on?
Which directory was it run in?
Which directory contains the output files?
What was the process id (PID) of the running bali-phy program?
What random seed was used?
What was the input file?
What alphabet was used to read in the file?
What substitution model was used to analyze the file?
What insertion/deletion model was used to analyze the file?
</p><pre class="screen"><code class="prompt">%</code> <strong class="userinput"><code>less 25-muscle-1/C1.out</code></strong>
</pre><p>Examine the file containing the sampled trees:
</p><pre class="screen"><code class="prompt">%</code> <strong class="userinput"><code>less 25-muscle-1/C1.trees</code></strong>
</pre><p>Examine the file containing the sampled alignments:
</p><pre class="screen"><code class="prompt">%</code> <strong class="userinput"><code>less 25-muscle-1/C1.P1.fastas</code></strong>
</pre><p>Examine the file containing the successive best alignment/tree pairs visited:
</p><pre class="screen"><code class="prompt">%</code> <strong class="userinput"><code>less 25-muscle-1/C1.MAP</code></strong>
</pre>
</div>

<div class="section"><div class="titlepage"><div><div><h3 class="title"><a name="idp53068560"></a>4.2. Summarizing the output</h3></div></div></div>
<p>
Try summarizing the sampled numerical parameters (e.g. not trees and alignments):
</p><pre class="screen"><code class="prompt">%</code> <strong class="userinput"><code>statreport --help</code></strong>
<code class="prompt">%</code> <strong class="userinput"><code>statreport 25-muscle-1/C1.p 25-muscle-2/C1.p &gt; Report</code></strong>
<code class="prompt">%</code> <strong class="userinput"><code>statreport 25-muscle-1/C1.p 25-muscle-2/C1.p --mean &gt; Report</code></strong>
<code class="prompt">%</code> <strong class="userinput"><code>statreport 25-muscle-1/C1.p 25-muscle-2/C1.p --mean --median &gt; Report</code></strong>
<code class="prompt">%</code> <strong class="userinput"><code>less Report</code></strong>
</pre><p>Now lets examine the summaries using a graphical program.  If you are using Windows or Mac, run Tracer, and press the <span class="guilabel">+</span> button to add these files.  What kind of ESS do you get?  If you are using Linux, do 
</p><pre class="screen"><code class="prompt">%</code> <strong class="userinput"><code>tracer 25-muscle-1/C1.p 25-muscle-2/C1.p &amp;</code></strong>
</pre><p>Now lets compute the consensus tree for these two runs:
</p><pre class="screen"><code class="prompt">%</code> <strong class="userinput"><code>trees-consensus --help</code></strong>
<code class="prompt">%</code> <strong class="userinput"><code>trees-consensus 25-muscle-1/C1.trees 25-muscle-2/C1.trees &gt; c50.PP.tree</code></strong>
<code class="prompt">%</code> <strong class="userinput"><code>trees-consensus 25-muscle-1/C1.trees 25-muscle-2/C1.trees --report=consensus &gt; c50.PP.tree</code></strong>
<code class="prompt">%</code> <strong class="userinput"><code>less consensus</code></strong>
<code class="prompt">%</code> <strong class="userinput"><code>figtree c50.PP.tree &amp;</code></strong>
</pre><p>Now lets see if there is evidence that the two runs have not converged yet:
</p><pre class="screen"><code class="prompt">%</code> <strong class="userinput"><code>trees-bootstrap --help</code></strong>
<code class="prompt">%</code> <strong class="userinput"><code>trees-bootstrap 25-muscle-1/C1.trees 25-muscle-2/C1.trees &gt; partitions.bs</code></strong>
<code class="prompt">%</code> <strong class="userinput"><code>less partitions.bs</code></strong>
</pre><p>
Now lets use the analysis script to run all the summaries and make a report:
</p><pre class="screen"><code class="prompt">%</code> <strong class="userinput"><code>bp-analyze.pl 25-muscle-1/ 25-muscle-2/</code></strong>
<code class="prompt">%</code> <strong class="userinput"><code>firefox Results/index.html &amp;</code></strong>
</pre>
</div>
</div>

<div class="section"><div class="titlepage"><div><div><h2 class="title" style="clear: both"><a name="idp53085376"></a>5. Multi-gene analyses</h2></div></div></div>
<p>This is described in more detail in section 4.3 of the <a class="ulink" href="http://www.bali-phy.org/README.html" target="_top">manual</a>.</p>
<p>
Download HIV env and pol alignments with the same sequence names:
</p><pre class="screen"><code class="prompt">%</code> <strong class="userinput"><code>wget http://nucleus.biology.duke.edu/~bredelings/alignment_files/env-common.fasta</code></strong>
<code class="prompt">%</code> <strong class="userinput"><code>wget http://nucleus.biology.duke.edu/~bredelings/alignment_files/pol-common.fasta</code></strong>
</pre><p>Try to analyze these genes jointly under a codon model:
</p><pre class="screen"><code class="prompt">%</code> <strong class="userinput"><code>bali-phy --test --alphabet Codons env-common.fasta pol-common.fasta --smodel M0</code></strong>
</pre><p>Take a look at the lengths of these files, to see whether the sequencealignment lengths are multiples of 3:
</p><pre class="screen"><code class="prompt">%</code> <strong class="userinput"><code>alignment-info pol-common.fasta</code></strong>
<code class="prompt">%</code> <strong class="userinput"><code>alignment-info env-common.fasta</code></strong>
</pre><p>Now edit the fasta files so that the lengths are a multiple of three:
</p><pre class="screen"><code class="prompt">%</code> <strong class="userinput"><code>alignment-cat -c1-951 env-common.fasta &gt; env-common2.fasta</code></strong>
<code class="prompt">%</code> <strong class="userinput"><code>alignment-cat -c1-1068 pol-common.fasta &gt; pol-common2.fasta</code></strong>
</pre><p>Try running bali-phy again:
</p><pre class="screen"><code class="prompt">%</code> <strong class="userinput"><code>bali-phy --test --alphabet Codons env-common2.fasta pol-common2.fasta --smodel M0</code></strong>
</pre><p>Now open the fasta files in an editor and replace all nucleotide ambiguity codes K and M with N.  This time the run should succeed:
</p><pre class="screen"><code class="prompt">%</code> <strong class="userinput"><code>bali-phy --test --alphabet Codons env-common2.fasta pol-common2.fasta --smodel M0</code></strong>
</pre><p>By default, the M0 substitution model will be selected for both genes.  However, each gene will have a separate instance of the model with its own gene-specific values for the model parameters.  Try specifying that both genes share one instance of the model parameters:
</p><pre class="screen"><code class="prompt">%</code> <strong class="userinput"><code>bali-phy --iterations=10 --alphabet Codons env-common2.fasta pol-common2.fasta --smodel=1,2:M0</code></strong>
</pre><p>Now try reading one gene as Codons, and the other as DNA:
</p><pre class="screen"><code class="prompt">%</code> <strong class="userinput"><code>bali-phy --iterations=10 --alphabet=1:Codons --alphabet=2:DNA env-common2.fasta pol-common2.fasta</code></strong>
</pre>
</div>

</div></body></html>
